## April 
### 20190405
Still no word from Peter Ralph. 

Simulated 250K with normal SLiM on cluster and laptop (~ 6Hrs w/ 200Gb on cluster, ~13 Hrs on laptop).

Potential positive bias of neutral variant, more detail here: https://docs.google.com/document/d/1FxlW1e1SDvzwcfRGqR-iTOex_paoQQID-WbZzyKJzwE/edit#

### 20190408
Comparing the effect of different MAF filtering levels on the GF output.

Below is a visual of all neutral allele frequencies from the 250K simulation just prior to the environmental shift. Colors of each block correspond to X-location:

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/All_AF_hist.png)

Next here is a histogram of neutral allele frequencies right before env. shift including only those alleles which were flagged as R2>0 by GF:

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/R2_AF_unFilt.png)

Then thereâ€™s a histogram of neutral allele frequencies right before env. shift showing only those alleles which were flagged as R2>0 after filtering for MAF > 0.01

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/R2_AF_Filt_0.01.png)

In the unfiltered data the neutral difference between the allele frequencies after-before environmental shift seems like what you'd expect from truly neutral markers

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/AllNeutral_alleleFreq.png)

However, when only looking at alleles which have been flagged by GF (r^2 >0) the differences are exclusively positive (only those neutral alleles which increased in frequency are included)

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/R2Neutral_alleleFreq.png)

Filtering for MAF < 0.01 resulted in the signal flipping, and now the difference in allele frequency of those alleles which were flagged by GF (r^2 >0) is exclusively negative (only those neutral alleles which decreased in frequency are included)

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/R2_Diff_AF_Filt_0.01.png)

If GF is arbitrarily picking a positive or negative association between an env. variable and allele frequencies, how come the M2 allele is persistently included despite having a non-linear associated allele frequency across the environmental gradient? Need to think on this some more.

### 20190409 KEL
Thoughts: Gradient Forests should be agnostic to the direction of *future* allele frequency change, but it should also be agnostic to the *sign of the correlation between allele frequency and environment*. The histograms are plotting each allele 10 times (once in each population) so it would be better to plot the overall MAF in the metapopulation to correspond with the overall allele frequency change that is being measured. The histograms still do not compare loci with R^2 = 0 to R^2 > 0. Let's lay out a plan for specific plots in our meeting today.

Also, I would recommend organizing the Notebook to have an /img folder, and using individual posts for each day or monthly journals. I would also add a README that indicates where previous notes and other relevant documents may be found.

### 20190412
Comparing the allele frequencies of the VCF files and the tree sequence recording (TSR).

All allele ferquencies taken from Generation 250,000 of the SLiM run (pre-environment shift). Nice bi-phasic distribution of near-zero alleles and near-fixed alleles.

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/Alelle_freq_VCF_G250K.png)

All allele ferquencies taken from Generation 250,300 of the SLiM run (post-environment shift). Again, a nice bi-phasic distribution of near-zero alleles and near-fixed alleles. Looks very similar to the 250K generation histogram but is actually slightly different.

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/Alelle_freq_VCF_G250.3K.png)

Finally the allele frequencies from the tree sequence recorded (TSR) file, recapitation was performed with Pyslim, and neutral mutations added with msprime. There are no fixed alleles in this output. This maybe something to do with the way msprime's _mutate_ function opperates.

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/Alelle_freq_TSR_G250.3K.png)

This TSR file from the SLiM simulation (so post env-shift) was recapitated and mutations added with the code below, using the following:

```recap = ts.recapitate(recombination_rate = 1e-06, migration_matrix=mig_mat, Ne=100, random_seed=1)```
```mutated = msprime.mutate(recap, rate=1e-7, random_seed=741, keep=True)```

### 20190415
Repeated the addition of mutations on the recapitated tree file from a new simulation run (TTT_SP_100_linear_EnvShift_10RC_TSR.slim) via msprime's mutate function. No fixed neutral mutations.
This is starting to make sense to me. There's no reason to map on fixed mutations since that'll be purged anyway in most analyses. The only way to be able to observe fixed mutations then is to simulate variants that reach fixation over the duration of the run. Until we figure out a way to read a recapitated and mutated file back into SLiM to finish the analysis, working with the 250K G. analyses is the only way to see neutral mutations that reach fixation. 

### 20190418
R code now takes in VCF output files and reformats for use in GF analysis. First run of GF with all 250K generations worth of allele variants results in output below:

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/GF_output_allR2.png)

Clearly there are way more false positive R^2>0 variants when sampling everything (and not just 1000 random alleles). 

Out of 31,675 M1 variants, 8,117 were flagged as having an R^2>0. That's 25.62589% of all the neutral variants. Should be noted that that's including all variants; no MAF filtering. More on that below.

Visualization of the cumulative importance curve, with demes mapped on by environment (A). Expected future environment represented by A circumflex).

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/GF.png)

### 20190419
Comparing histograms of allele frequencies flagged by GF as R^2 > 0 and those with R^2 = 0. These are prior to MAF filtering. Without proper filtering some 25.6% of all variant alleles were flagged as significant ( R^2 > 0) by GF

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/AF_R2_G0.png)

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/AF_R2_0.png)

Filtering for MAF < 0.05 and > 0.99 (going from 31,636 variants to 6,198), we see a substantial increase in the proportion flagged as significant by GF, with 3,366 alleles out of 6,198, or 54.3%, being used. Re-running GF with the filtered data set we get the allele frequency histograms below:

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/AFfilt_R2_G0.png)

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/AFfilt_R2_0.png)

Plotting the cumulative importance of the filtered dataset shows pretty much the same pattern as pre-filtration. This time the alleles that are linked (are in the same linkage group as the M2 allele) are coloured in cyan, and unlinked loci are coloured in orange. 15.7% of variants (530 R^2>0 alleles) were linked, and 84.25% (2,836 R^2>0 alleles) were unlinked. This is a slight increase from the unfiltered data set where 13% (1,055 out of all 8,118 R^2>0 alleles) were linked and 87% (7,063) were unlinked. 

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/GF_filt_col.png)

Linkage definitely contributes to the R^2 signal. So even though a majority of R^2>0 alllels are unlinked, those with the highest cumulative importance scores are linked to M2. 

Comparing the Spearman's correlation coefficient between allele frequency (filtered) of each variant to it's environment and it's R^2 value assigned to it by GF. The red dot is the M2 allele:

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/RHOvsR2.png)

This suggests that there are a number of linked alleles that are negatively correlated with the environment. This may be because they don't co-occurr with the M2 allele, so maybe playing around with the recombination rate would reduce the number of linked alleles that show an inverse relationship with the environment from the M2 allele.

Looking at the absolute value of rho shows how at moderate values of correlation R^2 is slightly more conservative in its estimation of importance than just a straight correlation of allele frequency to environment. However, at very low values of correlation R^2 actually seems to comparatively inflate the environmentl association.

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/RHOvsR2_abs.png)

### 20190422

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/CImp_Pop.png)
![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/M2_diff.png)
![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/M2diff_CImp.png)
![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/M1diff.png)
![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/M1diff_CImp.png)
![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/FST_diff.png)
![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/FSTdiff_CImp.png)
