## May 
### 20190501

Streamlining output of 250K SLiM simulation.
Summary stats: 
#### One per sim
% of alleles with R^2>0 \
% of R^2>0 unlinked to M2 \
F_ST_landscape_before \
F_ST_landscape_after \
Diff summaries of above tables per sim \
Correlation(rho and p-value) across loci_envrionment (Rho) and R^2 of locus \
Correlation across F_ST_locus_landscape_before and R^2 \
Correlation across F_ST_genome_ba_perPop and Diff_C.Imp \
Correlation across Rel_Fit_diff and Diff_C.Imp \
Correlation across M2_AF_diff_perPop and Diff_C.Imp \
Correlation across F_ST_M2_diff_perPop and Diff_C.Imp 


Altering parameters to include in output: \
Migration rate \
Recombination rate \
Mutation rate \
Pop size \
Env. rate \
Env. time points \
Sim type nickname (quick descriptor for more complex changes of parameters)


#### One per loci per sim 
Spearman’s correlation coefficient between AF of each variant to it’s environment and it’s R^2 value \
Linked/unlinked status \
Physical position \
Linkage group \
Locus type \
MAF/Het \
F_ST_locus_landscape_before \
F_ST_locus_landscape_after 

#### One per population per sim 
Pop ID (1-100) \
X-location \
Y-location \
GF_env \
Act_env \
Diff_GF_Act \
Extrapolation_logical \
C.Imp_current \
C.Imp_future \
Delta_C.Imp \
M2_AF_before \
M2_AF_after \
M1_AF_before \
M1_AF_after \
F_ST_genome_ba_perPop \
F_ST_M2_ba_perPop \
Rel_Fit_before \
Rel_Fit_after \
Rel_Fit_diff 

Generate meta data README file to explain columns.


Comparing the number of generations run to the coalescence of the generation. Image below represents the average number of roots across all trees in a TSR file. Once all the trees have one root we'll consider the run fully coalesced.

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/Coal_Gen.png)

Hopefully it does actually coalesce close to 250K, but will try a quick spot check to see if mutation mapping is consistent across generations if seed values are kept the same.

Comparison of allele frequency spectrums between VCF from 250K generation SLiM run, and Tree Sequence Recording (TSR) output that's been recapitated and had neutral mutations added via Pyslim:

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/VCF_SLiM_comp.png)

### 20190501

Looking at the number of loci which have coalesced by Generation 250K, we see that 99.99% of lineages in the TSR file have coalesced at that point (leaving just 17 of the 169,032 lineages with more than one root). Once every lineage in the TSR file has only one root all lineages would be considered to be coalesced. The average number of roots observed in the file at Gen. 250K is 1.00010057267263

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/VCF_SLiM_comp2.png)
![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/VCF_SLiM_comp2_log.png)

Considering a very simplistic formula for the expected time to coalescence for 200,000 individuals (E[tMRCA] = sum(2/(k\*(k-1)))\*200000)) we get 399,998. Generally we would always expect coalescence time to be roughly 2\*Ne so with 200,000 individuals we should expect the coalescence at the MRCA by Generation 400,000. This should be a conservative (_i.e._ longer) estimated time than will probably be true since we're estimating coalsecence at a number of linked sites, and reasonably low recombination rate (1e-7).

#### Estimating mutation overlay on two sequential TSR files

A single run of SLiM recipe 16.1 ("A minimal tree-seq model") was performed with a .tree file output at Generation 5000 and again at Generation 5300. Each tree file (file name _ts_ below) was recapitated and had neutral mutations overlayed with the msprime _mutate_ command \
``recap = ts.recapitate(recombination_rate = 1e-06, random_seed=1)`` \
``mutated = msprime.mutate(recap, rate=1e-7, random_seed=1, keep=True)`` \
and then read out as a VCF file with the command \
``with open("G5000.vcf", "w") as vcf_file:
    mutated.write_vcf(vcf_file, 2)`` \
These files where then read into R and the change in allele frequency and position analysed with package vcfR.

The G5K file had 147,699 variants, while G5.3K 152,772. 
Since the two outputs are 300 generations apart, the G5300 file had more variants (specifically 5,073 more). With a per unit of sequence length per generation mutation rate of 1e-7, a genome size of 100 million bps across 300 generations would be expected to yield 3000 new mutations. Across 500 diploid individuals that would then be 30,000 **(this can't be right, but why wouldn't the number of genomes in the simulation inflate the number of new variants seen?)**

In the msprime manual it's explicitly stated that "If the same seed is specified and all other parameters are equal then the same mutations will be generated." The caveat here seems to be that generation time is a distinct parameter. If the same mutations were being generated, we would generally expect that most of variants in the two files would share _roughly_ the same physical position (since there is an issue with the way msprime maps infinite site model mutations onto a limited sized genome the positions are not 100% going to match. However when comparing the number of positions shared between the two VCF files alarmingly few positions are actually shared. Only 322 exact positions out of the whole set were shared between the two files, and including variants plus or minus 1 position away, only 777 variants were shared. This suggests that at most only 0.5% (777/147699) of the variants are actually shared across the two files separated by 300 generations. If we expand the positions to include all those plus or minus 2 away from each other to be potentially shared we only see that number go up to 0.8% (1225/147699). 




