## May 
### 20190501

Streamlining output of 250K SLiM simulation.
Summary stats: 
#### One per sim
% of alleles with R^2>0 \
% of R^2>0 unlinked to M2 \
F_ST_landscape_before \
F_ST_landscape_after \
Diff summaries of above tables per sim \
Correlation(rho and p-value) across loci_envrionment (Rho) and R^2 of locus \
Correlation across F_ST_locus_landscape_before and R^2 \
Correlation across F_ST_genome_ba_perPop and Diff_C.Imp \
Correlation across Rel_Fit_diff and Diff_C.Imp \
Correlation across M2_AF_diff_perPop and Diff_C.Imp \
Correlation across F_ST_M2_diff_perPop and Diff_C.Imp 


Altering parameters to include in output: \
Migration rate \
Recombination rate \
Mutation rate \
Pop size \
Env. rate \
Env. time points \
Sim type nickname (quick descriptor for more complex changes of parameters)


#### One per loci per sim 
Spearman’s correlation coefficient between AF of each variant to it’s environment and it’s R^2 value \
Linked/unlinked status \
Physical position \
Linkage group \
Locus type \
MAF/Het \
F_ST_locus_landscape_before \
F_ST_locus_landscape_after 

#### One per population per sim 
Pop ID (1-100) \
X-location \
Y-location \
GF_env \
Act_env \
Diff_GF_Act \
Extrapolation_logical \
C.Imp_current \
C.Imp_future \
Delta_C.Imp \
M2_AF_before \
M2_AF_after \
M1_AF_before \
M1_AF_after \
F_ST_genome_ba_perPop \
F_ST_M2_ba_perPop \
Rel_Fit_before \
Rel_Fit_after \
Rel_Fit_diff 

Generate meta data README file to explain columns.


Comparing the number of generations run to the coalescence of the generation. Image below represents the average number of roots across all trees in a TSR file. Once all the trees have one root we'll consider the run fully coalesced.

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/Coal_Gen.png)

Hopefully it does actually coalesce close to 250K, but will try a quick spot check to see if mutation mapping is consistent across generations if seed values are kept the same.

Comparison of allele frequency spectrums between VCF from 250K generation SLiM run, and Tree Sequence Recording (TSR) output that's been recapitated and had neutral mutations added via Pyslim:

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/VCF_SLiM_comp.png)

### 20190501

Looking at the number of loci which have coalesced by Generation 250K, we see that 99.99% of lineages in the TSR file have coalesced at that point (leaving just 17 of the 169,032 lineages with more than one root). Once every lineage in the TSR file has only one root all lineages would be considered to be coalesced. The average number of roots observed in the file at Gen. 250K is 1.00010057267263

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/VCF_SLiM_comp2.png)
![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/VCF_SLiM_comp2_log.png)

Considering a very simplistic formula for the expected time to coalescence for 100,000 individuals (E[tMRCA] = sum(2/(k\*(k-1)))\*200000)) we get 399,996. Generally we would always expect coalescence time to be roughly 2\*Ne so with 100,000 diploid individuals we should expect the coalescence at the MRCA by Generation 400,000. This should be a conservative (_i.e._ longer) estimated time than will probably be true since we're estimating coalsecence at a number of linked sites, and reasonably low recombination rate (1e-7).

#### Estimating mutation overlay on two sequential TSR files

A single run of SLiM recipe 16.1 ("A minimal tree-seq model") was performed with a .tree file output at Generation 5000 and again at Generation 5300. Each tree file (file name _ts_ below) was recapitated and had neutral mutations overlayed with the msprime _mutate_ command \
``recap = ts.recapitate(recombination_rate = 1e-06, random_seed=1)`` \
``mutated = msprime.mutate(recap, rate=1e-7, random_seed=1, keep=True)`` \
and then read out as a VCF file with the command \
``with open("G5000.vcf", "w") as vcf_file:
    mutated.write_vcf(vcf_file, 2)`` \
These files where then read into R and the change in allele frequency and position analysed with package vcfR.

The G5K file had 147,699 variants, while G5.3K 152,772. 
Since the two outputs are 300 generations apart, the G5300 file had more variants (specifically 5,073 more). With a per unit of sequence length per generation mutation rate of 1e-7, a genome size of 100 million bps across 300 generations would be expected to yield 3000 new mutations. Across 500 diploid individuals that would then be 30,000 **(this can't be right, but why wouldn't the number of genomes in the simulation inflate the number of new variants seen?)**

In the msprime manual it's explicitly stated that "If the same seed is specified and all other parameters are equal then the same mutations will be generated." The caveat here seems to be that generation time is a distinct parameter. If the same mutations were being generated, we would generally expect that most of variants in the two files would share _roughly_ the same physical position (since there is an issue with the way msprime maps infinite site model mutations onto a limited sized genome the positions are not 100% going to match. However when comparing the number of positions shared between the two VCF files alarmingly few positions are actually shared. Only 322 exact positions out of the whole set were shared between the two files, and including variants plus or minus 1 position away, only 777 variants were shared. This suggests that at most only 0.5% (777/147699) of the variants are actually shared across the two files separated by 300 generations. If we expand the positions to include all those plus or minus 2 away from each other to be potentially shared we only see that number go up to 0.8% (1225/147699). 


### ~~20190517~~
### 20190524
### Post talk with Matt

#### Reworked summary tables
#### One per sim
"No.A" - Total number of alleles \
"PR2" - Percentage of alleles with R^2>0 \
"PR2_UL" - Percentage of R^2>0 unlinked to M2 \
"F_ST_lb" - F_ST across landscape before environmental shift (averaged across all loci) \
"F_ST_la" - F_ST across landscape after environmental shift (averaged across all loci) \
"Rel.Fit_lb" - Relative fitness across landscape before env. shift (is this an average across all populations?) \
"Rel.Fit_la" - Relative fitness across landscape after env. shift (is this an average across all populations?) \
"Rho_EnvR2" - Spearman’s correlation coefficient (Rho) between (i) Rho of allele frequency and environment and (ii) R^2 of allele \
"Rho_EnvR2_p" - p-value of above \
"Rho_FST_R2" - Rho between (i) allele specific F_ST across landscape before environmental shift and (ii) R^2 of allele \
"Rho_FST_R2_p" - p-value of above \
"Rho_FST_DC.Imp" - Rho between (i) genome-wide F_ST value of pop. at pre-shifted environment and same pop. post shift (true genetic offset for genome) and (ii) the difference in cumulative importance across all loci (with R^2 > 0) estimated at pre and post env. shift (GF predicted genomic offset) \
"Rho_FST_DC.Imp_p" - p-value of above \
"Rho_M2FST_DC.Imp" - Rho between (i) M2 allele F_ST value of pop. at pre-shifted environment and same pop. post shift (true genetic offset for M2) and (ii) the difference in cumulative importance for M2 estimated at pre and post env. shift (GF predicted genomic offset for adaptive allele) \
"Rho_M2FST_DC.Imp_p" - p-value of above \
"Rho_RelFitD_DC.Imp" - Rho between (i) the difference in relative fitness of each population pre and post environment shift (true genomic vulnerability for fitness) and (ii) the difference in cumulative importance for M2 estimated at pre and post env. shift \
"Rho_RelFitD_DC.Imp_p" - p-value of above \
"Rho_DM2_DC.Imp" - Rho between (i) the difference in allele frequency of selected allele M2 in pop. pre and post environment shift (true genetic offset for M2) and (ii) the difference in cumulative importance for M2 estimated at pre and post env. shift (GF predicted genomic offset for adaptive allele)\
"Rho_DM2_DC.Imp_p" - p-value of above 


#### One per allele per sim 
"R2" - Locus R^2 value from GF \
"Rho" - Rho between the Rho of allele frequency to environment and R^2 value \
"Link" - Is allele linked or unlinked to allele(s) under selection  **(Do you have some text describing how you decided this that I can look over?)** \
"PP" - Physical position of allele in genome **(Recommend POS to be consistent with genomic datasets and VCF files)** \
"LG" - Which linkage group contains the allele \
"LT" - "Locus Type" Is the allele neutral "M1" or under selection "M2" \
"MAF" - Minor Allele Frequency of the allele (across whole metapopulation) \
"Het" - Heterozygosity of the allele (across whole metapopulation) \
"F_ST_llb" - Allele specific F_ST across landscape before environmetal shift \
"F_ST_lla" - Allele specific F_ST across landscape after environmental shift 

Question - will this table only contain alleles with MAF > 0.05? \
    ÁJL - I was thinking that it only contain the filtered alleles (MAF > 0.05 & < 0.95)

#### One per population per sim 
"Pop" - Pop ID (1-100) \
"X" - X-location of pop. in metapopulation matrix \
"Y" - Y-location of pop. in metapopulation matrix \
"Env_before" - Environmental value at pop. location before environmental shift \
"Env_after" - Environmental value at pop. location at the end of environmental shift \
"Diff_env" - Difference in environment value at start and at end of environmental shift \
"Env_range" - Is the environmental value (rounded to 10th decimal place) considered by GF to calc cumulative importance present both before and after the env. shift \
"C.Imp_genome_before" - Cumulative importance estimated genome-wide at pre-shifted environment values \
"C.Imp_genome_after" - Cumulative importance estimated genome-wide at post-shifted environment values \
"D_C.Imp_genome" - Difference in cumulative importance estimated genome-wide at pre and post environment shift **GF Genetic Offset for genome** \
"C.Imp_M2_before" - Cumulative importance estimated for only M2 at pre-shifted environment values \
"C.Imp_M2_after" - Cumulative importance estimated for only M2 at post-shifted environment values \
"D_C.Imp_M2" - Difference in cumulative importance for only M2 estimated at pre and post environment shift **GF Genetic offset for M2** \
"M2_AF_before" - Allele frequency of selected allele M2 in pop. pre environment shift \
"M2_AF_after" - Allele frequency of selected allele M2 in pop. post environment shift \
"M2_AF_diff" - Difference in allele frequency of selected allele M2 in pop. pre and post environment shift \
"M1_AF_before" - Average allele frequency of neutral alleles in pop. pre environment shift \
"M1_AF_after" - Average allele frequency of neutral alleles in pop. post environment shift \
"F_ST_genome_bef.aft." - Genome wide F_ST value comparing pop. at pre-shifted environment to same pop. post shift **True genetic offset for genome**\
"F_ST_M2_bef.aft" - M2 allele F_ST value comparing pop. at pre-shifted environment to same pop. post shift **True genetic offset for M2**\
"Rel_Fit_before" - Relative fitness of each population pre environment shift \
"Rel_Fit_after" - Relative fitness of each population post environment shift \
"Rel_Fit_diff" - Difference in relative fitness of each population pre and post environment shift **True fitness offset** 

### 20190525-28 ###
Confirmed that the difference seen in the per population M2 allele frequency between the SLiM calculated value and the VCF R code calculation was due to SLiM calculating it as an "early()" event, and the VCF file was being generated as a "late()" event. SLiM code 'TTT_SP_100_VCF_300K.slim' has been updated so that both outputs are performed "early()". This was done because the SLiM function "cachedFitness()", which is an integral part of the summary output file containing the SLiM calculated M2 allele frequency, must be called as an "early()" event.

An error came up when estimating "F_ST_M2_bef.aft" via the OutFLANK function "MakeDiploidFSTMat" in the event that a certain population contained no homozygotes of the M2 allele. This was fixed by changing the "MakeDiploidFSTMat" condition that factor levels in the SNP matrix must equal "012" or "0129" to allow for a condition in which levels can include any of the levels contained in "0129".

Summary files in the R code 'TTT_VCF_analysis_clean.R' should output three summary files: \
'Summary_Pop_<SEED>.csv' - Containing the "One per population per sim" summaries specified above **Complete** \
'Summary_Allele_<SEED>.csv' - Containing the "One per allele per sim" summaries specified above **Implementing 'unlinked' analysis - need to add description** \
'Summary_Sim_<SEED>.csv' - Containing the "One per sim" summaries specified above **Implementing "F_ST_lb" and "F_ST_la"** 
    
Discussed having at least 10 replicates in time for Evolution talk. Each SLiM run on the Defiance cluster takes little less than 40 hours real time. All ten can easily be run in parallel.

### 20190528 ###
Since the F_ST calculations as implemented actually re-make the G matrices, and currently filtration for MAF is done post per-Pop allele frequency calculation earlier in the code, the current FST values are being calculated on alleles unfiltered for MAF < 0.05 & > 0.95 \
Fixed. F_ST outputs are now from filtered alleles.
