We're attempting to generate a tree file in SLiM with a single allele under selection, and properly coalesced neutral alleles that we can compare across two distinct time points (before and after an environmental shift).
We want to have properly coalseced neutral mutations before the environmental shift begins, but it takes a long time running SLiM to get that (400,000+ generations).
Because we can't read a mutated recapitated tree file back into SLiM to continue the simulation, we were hoping to map identical mutations onto two tree file outputs made at each time point (T1 and T2).

When we output the T1 tree file and recapitate it, we can read it back into SLiM and at the second time point (T2) output a tree file that appears to have started from the same tree topology as T1.

Each tree file (file name _ts_ below) was recapitated and had neutral mutations overlayed with the msprime _mutate_ command using the same seed value for both T1 and T2 \
``recap = ts.recapitate(recombination_rate = 1e-06, random_seed=1)`` \
``mutated = msprime.mutate(recap, rate=1e-7, random_seed=1, keep=True)`` \
and then read out as a VCF file with the command \
``with open("G5000.vcf", "w") as vcf_file:
    mutated.write_vcf(vcf_file, 2)`` \
These files where then read into R and the change in allele frequency and position analysed with package vcfR.

However, when we add mutations to both trees and output them as VCF files, the mutations between T1 and T2 only share _roughly_ the same position at 0.8% of sites. 
This is accounting for the staggering that can happen when the files are read out to VCF format by lumping in all variants that share positions at +/- 2 base positions.

So is there another way to get the same properly coalesced neutral mutations across two different time points?
