## 2020 07 13

__Missing genotypes in vcf files__

After finding some NA allele frequencies in our downstream visualization code, I traced the NA allele frequencies back to blank genotypes "" in the VCF files read into R, that seemed to read random genotypes as blank each time the same file was read in:

```
#First trial
> vcf1 <- read.vcfR("test.vcf")
Scanning file to determine attributes.
...
Processed variant: 9423
All variants processed

> which(vcf1@gt[,-1]=="")
[1] 228810 229751 229780 234489


#Second trial
> vcf1 <- read.vcfR("test.vcf")
Scanning file to determine attributes.
...
Processed variant: 9423
All variants processed

> which(vcf1@gt[,-1]=="")
[1] 226878 228705 230215 232006 232873 232922


#Third trial
> vcf1 <- read.vcfR("test.vcf")
Scanning file to determine attributes.
...
Processed variant: 9423
All variants processed

> which(vcf1@gt[,-1]=="")
[1] 228555 229931 232006 232856


#Fourth trial
> vcf1 <- read.vcfR("test.vcf")
Scanning file to determine attributes.
...
Processed variant: 9423
All variants processed

> which(vcf1@gt[,-1]=="")
[1] 230321 232006


#Fifth trial
> vcf1 <- read.vcfR("test.vcf")
Scanning file to determine attributes.
...
Processed variant: 9423
All variants processed

> which(vcf1@gt[,-1]=="")
[1] 227752 229518 232182 232910 233457
```

At first it looked to me that the original VCF files were not the problem since none of them had missing genotypes prior to being read into R. However Katie noticed that the REF and ALT columns of the VCF files had occasional formatting errors:

![image](https://github.com/TestTheTests/TTT_Offset_Vulnerability_GF_Sims/blob/master/Notebook/img/SnapshotVCF.PNG)

The number that appears instead of the REF and ALT entries doesn't immediately seem to be related to the position data or to the individual IDs (either tskit or SLiM ID).

Example of the first 20 numbers and the position at which they show up in file T1_1594110774619_filtered_subset.recode.vcf:

|position	|number|
| :------------- | :----------: | 
|9927|417756|
|12908|486585|
|20080|595624|
|30484|636767|
|31644|452582|
|35127|474358|
|38076|364321|
|59118|475649|
|59152|461334|
|72745|608104|
|73191|623032|
|75540|627757|
|80690|640798|
|81110|529727|
|84822|373797|
|85427|528801|
|85858|567080|
|91141|423647|
|91163|345702|
|93959|219586|

These numbers show up in the original Pyslim output VCF files, and are carried over into the filtered and subset VCF files without returning any error messages. The absence of a REF character, and the presence of a six digit number in the ALT column doesn't appear to preclude vcftools from processing the original file. Only about 1% of sites show this error (~125/9000 sites).

This issue was previously brought up in February of this year, on the [Pyslim github](https://github.com/tskit-dev/pyslim/issues/65).

Since the root issue still does not have an elegant solution from the Pyslim developers, a quick and easy fix is a regex replacement<br/>
Find: \t\t[0-9]{1,6}
Replace: \t0\t1



   


