library(OutFLANK)  # outflank package
library(vcfR)
require(adegenet)
require(gdm)
require(gradientForest)
require(foreach)
require(doParallel)
require(OutFLANK)
require(pbapply)
require(gdata)
require(data.table)
require(PresenceAbsence)
require(ROCR)
require(modEvA)
require(ggplot2)
library(dplyr)
require(grid)
require(gridExtra)
require(gtools)
require(cowplot)
library(stringr)
require(reshape2)

setwd("/Users/akijarl/Desktop/PostDoc/SLiMstuff/SLiM_output/250K/VCF_output/")
#setwd("/Users/akijarl/Desktop/PostDoc/SLiMstuff/SLiM_output/")

vcf1 <- read.vcfR("TTT_SP_100_1599594485287_Gen250000.vcf", verbose = T)
vcf2 <- read.vcfR("TTT_SP_100_1599594485287_Gen250300.vcf", verbose = T)

vcfTSR <- read.vcfR("TTT_SP_100_linear_EnvShift_10RC_1649683388434_recap_mutate.vcf", verbose = T)
vcfTSR <- read.vcfR("TTT_SP_100_linear_EnvShift_10RC_1687770438261_recap_mutate.vcf", verbose = T)

### Convert VCF to 012 format ####
# Character matrix containing the genotypes
# individuals in columns
# Remove 1st column, which is 'Format'
geno1 <- vcf1@gt[,-1] 
geno2 <- vcf2@gt[,-1] 

genoTSR <- vcfTSR@gt[,-1] 

position1 <- getPOS(vcf1) # Positions in bp
chromosome1 <- getCHROM(vcf1) # Chromosome information
ID1 <- getINFO(vcf1) # Meta data information

position2 <- getPOS(vcf2) # Positions in bp
chromosome2 <- getCHROM(vcf2) # Chromosome information
ID2 <- getINFO(vcf2) # Meta data information

positionTSR <- getPOS(vcfTSR) # Positions in bp
chromosomeTSR <- getCHROM(vcfTSR) # Chromosome information
#IDTSR <- getINFO(vcfTSR) # Meta data information

#Create Genotype matrix
G1 <- matrix(NA, nrow = nrow(geno1), ncol = ncol(geno1))
G1[geno1 %in% c("0/0", "0|0")] <- 0
G1[geno1  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
G1[geno1 %in% c("1/1", "1|1")] <- 2

G2 <- matrix(NA, nrow = nrow(geno2), ncol = ncol(geno2))
G2[geno2 %in% c("0/0", "0|0")] <- 0
G2[geno2  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
G2[geno2 %in% c("1/1", "1|1")] <- 2

GTSR <- matrix(NA, nrow = nrow(genoTSR), ncol = ncol(genoTSR))
GTSR[genoTSR %in% c("0/0", "0|0")] <- 0
GTSR[genoTSR  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
GTSR[genoTSR %in% c("1/1", "1|1")] <- 2

#Check number of duplicate positions
sum(duplicated(position1))
sum(duplicated(position2))

sum(duplicated(positionTSR))

#Position of M2 allele
position1[grep("MT=2",ID1)]
position2[grep("MT=2",ID2)]

#Calculate allele frequencies
a_freq1 <- rowSums(G1)/(2*ncol(G1))
a_freq2 <- rowSums(G2)/(2*ncol(G2))
a_freqTSR <- rowSums(GTSR)/(2*ncol(GTSR))

#Visualize allele frequencies
hist(a_freq1, breaks=75,xlab="Allele frequency", main="Generation 250000, pre env. shift", col="cyan")
hist(a_freq2, breaks=75, xlab="Allele frequency", main="Generation 250300, post env. shift", col="orange")
hist(a_freqTSR, breaks=75, xlab="Allele frequency", main="From TSR file, Generation 250300, post env. shift", col="red")

plot(a_freq1[order(a_freq1)], ylab="Allele frequency", main="Generation 250000, pre env. shift", col="cyan")
plot(a_freq2[order(a_freq2)], ylab="Allele frequency", main="Generation 250300, post env. shift", col="orange")
plot(a_freqTSR[order(a_freqTSR)], ylab="Allele frequency", main="From TSR file, Generation 250300, post env. shift", col="red")

#Combine allele frequency, genomic position, & meta data into one data frame
PreN<-data.frame(a_freq1,position1,ID1,stringsAsFactors = F)
PreN[grep("MT=2",PreN$ID1),]

#Extract SLiM assigned MIDs to keep track of mutations
pat <- "MID=([^;]+);"
MID_pre<-paste("M",str_match(PreN$ID1, pat)[,2],sep="")
M2_MID<-paste("M",str_match(PreN$ID1[grep("MT=2",ID1)],pat)[,2],sep="")

PostN<-data.frame(a_freq2,position2,ID2)
PostN[grep("MT=2",PostN$ID2),]

MID_post<-paste("M",str_match(PostN$ID2, pat)[,2],sep="")

Start<-seq(1,10000,100)
Stop<-seq(100,10000,100)

Pop_afreq1<-NULL
for(i in 1:100){
  Pop_afreq1<-rbind(Pop_afreq1,rowSums(G1[,Start[i]:Stop[i]])/(2*ncol(G1[,Start[i]:Stop[i]])))
}

Pop_afreq1<-data.frame(Pop_afreq1)
colnames(Pop_afreq1)<-MID_pre

Pop_afreq2<-NULL
for(i in 1:100){
  Pop_afreq2<-rbind(Pop_afreq2,rowSums(G2[,Start[i]:Stop[i]])/(2*ncol(G2[,Start[i]:Stop[i]])))
}

Pop_afreq2<-data.frame(Pop_afreq2)
colnames(Pop_afreq2)<-MID_post

Pops <- NULL
for(j in 1:10){
  for(i in 1:10){
    Pops <- c(Pops,print(paste("A",i,sep="")))
  }
}

Pop_afreq1$Pops<-Pops
Pop_afreq1$Pops <- factor(Pop_afreq1$Pops, levels = Pop_afreq1$Pops[1:10])

Pop_afreq2$Pops<-Pops
Pop_afreq2$Pops <- factor(Pop_afreq2$Pops, levels = Pop_afreq2$Pops[1:10])

M2<-data.frame(Pop_afreq1[,c(which(colnames(Pop_afreq1)==M2_MID))])
colnames(M2)<-M2_MID

M2_2<-data.frame(Pop_afreq2[,c(which(colnames(Pop_afreq2)==M2_MID))])
colnames(M2_2)<-M2_MID

ggplot(data = M2)+
  geom_boxplot(aes(y=M2[,1],x=Pop_afreq1$Pops))

ggplot(data = M2_2)+
  geom_boxplot(aes(y=M2_2[,1],x=Pop_afreq2$Pops))

afreq1_N<-Pop_afreq1[,-c(which(colnames(Pop_afreq1)==M2_MID),length(Pop_afreq1))]

afreq2_N<-Pop_afreq2[,-c(which(colnames(Pop_afreq2)==M2_MID),length(Pop_afreq2))]

#Input Fit & Env file here
names(envPop) <- "envSelect"

alFreq<-cbind(afreq1_N,M2) #Keep track of which M2 generation is beign compared
#Filter for MAF<0.05 and >0.99
#alFreq<-aFreq[,]

ffreq250000<-freq250000[,which(colMeans(freq250000)>0)]

# number of cores to use for parallel processing
cores <- 3

#####
# build output GF data frames
gfR2tab <- function(gfMods.list, alFreqs){
  i=1
  while(is.null(gfMods.list[[i]])){i=i+1}
  tab <- do.call(rbind, gfMods.list)
  vrNm <- rep(row.names(tab)[1:nrow(gfMods.list[[i]])], 
              nrow(tab)/nrow(gfMods.list[[i]]))
  tab <- data.frame(variable=vrNm, tab)
  tab <- dcast(tab, SNPnames~variable, value.var="imps")
  envR2 <- rowSums(data.frame(tab[,-1]))
  R2Tab <- data.frame(tab, envR2=envR2)
  
  # get name of SNP if it has a positive R2
  posR2 <- unlist(lapply(gfMods.list, function(x){
    return(as.character(unique(x[,2])))}))
  
  # Find which loci have R2 < 0 (no GF model for those) & assign R2=0
  negR2 <- !(colnames(alFreqs) %in% posR2)
  negR2 <- colnames(alFreqs)[negR2] 
  
  noGF <- data.frame(matrix(0, nrow=length(negR2), ncol=ncol(R2Tab)))
  colnames(noGF) <- colnames(R2Tab)           
  noGF$SNPnames <- negR2
  
  R2Tab <- rbind(R2Tab, noGF)
  return(R2Tab[order(R2Tab$SNPnames),])}

##############################################
# Chunk to fit GF models to minor allele frequencies at the level of
# populations
# GF is fit to each SNP individually to 
# ease computational / memory burden

cl <- makeCluster(cores)
registerDoParallel(cl)

# use for each to run in parallel - each SNP modeled by GF on different processor
gfAllele.freq <- foreach(k=1:ncol(alFreq), .verbose=F, .packages=c("gradientForest", "data.table")) %dopar% {
  # get locus k & name with SNP ID
  locus <- data.frame(alFreq[,k])
  names(locus) <- colnames(alFreq)[k]
  
  # here's the GF modeling function
  gfLocus <- gradientForest(data=data.frame(envPop, locus),
                            #gfLocus <- gradientForest(data=data.frame(locPop, locus),
                            predictor.vars=colnames(envPop), 
                            #predictor.vars=colnames(locPop),
                            response.vars=colnames(alFreq)[k], 
                            corr.threshold=0.5, 
                            ntree=500, 
                            trace=F)
  
  # get result is the model is not = NULL (no fit)
  if(!is.null(gfLocus)){
    # cumulative importance values (to plot turnover functions)
    cImp <- cumimp(gfLocus, "envSelect", type="Species")
    cImp <- data.frame(rbindlist(cImp, idcol="allele"))
    # return data frame with allele ID, cumulative importance results,
    # and R2 value for GF model
    data.frame(cImp, r2=gfLocus$result)
  }
}
stopCluster(cl)

# Prep table of importance values for each SNP (rows) 
# and each var (columns)
gfAF <- lapply(gfAllele.freq, function(x){
  if(!is.null(x)){
    ttt <- x[1,c("r2", "allele")]
    return(data.frame(imps=ttt$r2, SNPnames=ttt$allele, row.names = "envSelect",
                      stringsAsFactors=F))}})

# run function to convert to table # then write to file
gfAllele.R2.af <- gfR2tab(gfAF, alFreq)
gfAllele.R2.af <- gfAllele.R2.af[match(mixedsort(gfAllele.R2.af$SNPnames), 
                                       gfAllele.R2.af$SNPnames),]
gfAllele.R2.af$SNPnames <- gsub("S", "", gfAllele.R2.af$SNPnames)

finalOut <- data.frame(SNPnames=gfAllele.R2.af$SNPnames, 
                       #cImp.envSelect.alleleFreq=gfAllele.R2.af$envSelect,
                       r_squared.alleleFreq=gfAllele.R2.af$envR2)

write.table(finalOut, paste("alFreq_",substr(plotTitle, start=19, stop=31),"Gen250000_gradientforests.Cpval",sep=""), row.names=F)
save(gfAllele.freq, file=paste("alFreq_",substr(plotTitle, start=19, stop=31),"Gen250000_gradientforests.Rdata",sep=""))

impDatList <- gfAllele.freq[unlist(lapply(gfAllele.freq, function(x){!is.null(x)}))]

impDat <- do.call(rbind, impDatList)

cImpMAF.sel<-impDat[which(impDat$allele==M2_MID),]

p.imp1 <- 
  ggplot() + 
  #geom_line(data=impDat, aes(x=x, y=y, group=allele), colour=rgb(0,0,0,0.4), lwd=0.5) +
  #facet_grid(. ~ strSel) +
  labs(y="Cumulative Importance", x="Environment") +
  geom_line(data=cImpMAF.sel, aes(x=x, y=y),
            colour=rgb(1,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  ggtitle("") +
  ggtitle(paste(plotTitle,"Gen250000")) +
  theme(plot.title = element_text(size=14, face="bold.italic"))

p.imp1

ggCand1<-cImpMAF.sel

(ggCand_n1_pre<-ggCand1[1,c("x","y")])
(ggCand_n2_pre<-ggCand1[2,c("x","y")])
(ggCand_n3_pre<-ggCand1[3,c("x","y")])
(ggCand_n4_pre<-ggCand1[4,c("x","y")])
(ggCand_n5_pre<-ggCand1[5,c("x","y")])
(ggCand_n6_pre<-ggCand1[6,c("x","y")])
(ggCand_n7_pre<-ggCand1[7,c("x","y")])
(ggCand_n8_pre<-ggCand1[8,c("x","y")])
(ggCand_n9_pre<-ggCand1[9,c("x","y")])

######################################
#Extrapolation 
#x = environmental variable
#y = cumulative importance
#ymin + (y-ymin)*(x-xmin)/(xmax-xmin) = 0 
#y = ymin - ymin*(xmax-xmin)/(x-xmin)
#####################################
x <-c(5:7)
y <- min(ggCand1$y)-((min(ggCand1$y)*(max(ggCand1$x)-min(ggCand1$x)))/(x-min(ggCand1$x)))
y2<-max(ggCand1$y)+y

x1 <- -5
y1 <- min(ggCand1$y)-((min(ggCand1$y)*(max(ggCand1$x)-min(ggCand1$x)))/(x1-min(ggCand1$x)))
x3 <- x1+3

#Visualize cumulative importance curve with placeholders for demes (grouped by environment)
p.imp1+
  geom_point(aes(x=x1,y=y1,size=2),fill=NA,shape=21)+
  geom_point(aes(x=x3,y=ggCand_n3_pre$y,size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n1_pre$x,y=ggCand_n1_pre$y,size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n1_pre$x+3,y=ggCand1$y[4],size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n2_pre$x,y=ggCand_n2_pre$y,size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n2_pre$x+3,y=ggCand1$y[5],size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n3_pre$x,y=ggCand_n3_pre$y,size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n3_pre$x+3,y=ggCand1$y[6],size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n4_pre$x,y=ggCand_n4_pre$y,size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n4_pre$x+3,y=ggCand1$y[7],size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n5_pre$x,y=ggCand_n5_pre$y,size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n5_pre$x+3,y=ggCand1$y[8],size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n6_pre$x,y=ggCand_n6_pre$y,size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n6_pre$x+3,y=ggCand1$y[9],size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n7_pre$x,y=ggCand_n7_pre$y,size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n7_pre$x+3,y=y2[1],size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n8_pre$x,y=ggCand_n8_pre$y,size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n8_pre$x+3,y=y2[2],size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n9_pre$x,y=ggCand_n9_pre$y,size=2),fill=NA,shape=21)+
  geom_point(aes(x=ggCand_n9_pre$x+3,y=y2[3],size=2),fill=NA,shape=21)+
  geom_text(aes(x=x1,y=y1+0.1,label=paste("A[",1,"]")),parse = TRUE)+
  geom_text(aes(x=x3,y=ggCand_n3_pre$y-0.1,label=paste("hat(A)[",1,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n1_pre$x,y=ggCand_n1_pre$y+0.1,label=paste("A[",2,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n1_pre$x+3,y=ggCand1$y[4]-0.1,label=paste("hat(A)[",2,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n2_pre$x,y=ggCand_n2_pre$y+0.1,label=paste("A[",3,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n2_pre$x+3,y=ggCand1$y[5]-0.1,label=paste("hat(A)[",3,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n3_pre$x,y=ggCand_n3_pre$y+0.1,label=paste("A[",4,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n3_pre$x+3,y=ggCand1$y[6]-0.1,label=paste("hat(A)[",4,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n4_pre$x,y=ggCand_n4_pre$y+0.1,label=paste("A[",5,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n4_pre$x+3,y=ggCand1$y[7]-0.1,label=paste("hat(A)[",5,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n5_pre$x,y=ggCand_n5_pre$y+0.1,label=paste("A[",6,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n5_pre$x+3,y=ggCand1$y[8]-0.1,label=paste("hat(A)[",6,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n6_pre$x,y=ggCand_n6_pre$y+0.1,label=paste("A[",7,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n6_pre$x+3,y=ggCand1$y[9]-0.1,label=paste("hat(A)[",7,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n7_pre$x,y=ggCand_n7_pre$y+0.1,label=paste("A[",8,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n7_pre$x+3,y=y2[1]-0.1,label=paste("hat(A)[",8,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n8_pre$x,y=ggCand_n8_pre$y+0.1,label=paste("A[",9,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n8_pre$x+3,y=y2[2]-0.1,label=paste("hat(A)[",9,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n9_pre$x,y=ggCand_n9_pre$y+0.1,label=paste("A[",10,"]")),parse = TRUE)+
  geom_text(aes(x=ggCand_n9_pre$x+3,y=y2[3]-0.1,label=paste("hat(A)[",10,"]")),parse = TRUE)+
  scale_x_continuous(name="Environment", limits=c(-5,7), breaks=seq(-5,7,2))+
  theme(legend.position="none")

#Proportion of R^2>0 alleles out of total
length(unique(impDat$allele))/length(alFreq)

#Get alleles which had R^2>0 (R2) and R^2=0 (R0)
R2<-alFreq[,colnames(alFreq)%in%impDat$allele]
R0<-alFreq[,!colnames(alFreq)%in%impDat$allele]

R2M<-data.frame(colnames(R2),colMeans(R2))
R0M<- data.frame(colnames(R0),colMeans(R0))

ggplot(data = R2M)+#, aes(colour=Pops))+
  geom_histogram(aes(R2M$colMeans.R2.))+
  xlab("Allele Freq of R^2>0 variants")+ 
  ylab("")+
  #scale_color_manual(values=colfunc(10))+
  geom_hline(yintercept = 0,col="grey") +
  theme(legend.position="none")

ggplot(data = R0M)+#, aes(colour=Pops))+
  geom_histogram(aes(R0M$colMeans.R0.))+
  xlab("Allele Freq of R^2=0 variants")+ 
  ylab("")+
  #scale_color_manual(values=colfunc(10))+
  geom_hline(yintercept = 0,col="grey") +
  theme(legend.position="none")

#Difference in cumulative importance
(DcumImp<-data.frame(rbind(ggCand_n3_pre$y-y1,
                           ggCand1$y[4]-ggCand_n1_pre$y,
                           ggCand1$y[5]-ggCand_n2_pre$y,
                           ggCand1$y[6]-ggCand_n3_pre$y,
                           ggCand1$y[7]-ggCand_n4_pre$y,
                           ggCand1$y[8]-ggCand_n5_pre$y,
                           ggCand1$y[9]-ggCand_n6_pre$y,
                           y2[1]-ggCand_n7_pre$y,
                           y2[2]-ggCand_n8_pre$y,
                           y2[3]-ggCand_n9_pre$y)))

rownames(DcumImp)<-c("A1","A2","A3","A4","A5","A6","A7","A8","A9","A10")
colnames(DcumImp)<-"DCumImp"
DcumImp$Pops <- Pops[1:10] 
DcumImp$Pops <- factor(DcumImp$Pops, levels = Pops[1:10])
DcumImp$Extr <- c("Y","N","N","N","N","N","N","Y","Y","Y")

colfunc<-colorRampPalette(c("orange","cyan"))

ggplot(data=DcumImp,aes(colour=Pops, shape=Extr))+
  geom_point(aes(x=DcumImp$Pops, y=DcumImp$DCumImp),size=4)+
  xlab("Populations")+
  ylab(expression(paste("'Genomic vulnerability' (",Delta,"CumImp)",sep="")))+
  scale_color_manual(values=colfunc(10))

DM2<-M2_2-M2
DM2$Pops<-Pops
DM2$Pops <- factor(DM2$Pops, levels = DM2$Pops[1:10])

ggplot(data = DM2)+
  geom_boxplot(aes(y=DM2[,1],x=DM2$Pops))

ggplot(data = DM2, aes(colour=Pops))+
  geom_boxplot(aes(y=DM2$M249814835,x=DM2$Pops))+
  xlab("")+ 
  ylab("Results of M2 Allele Freq(future) - Allele Freq(present)")+
  scale_color_manual(values=colfunc(10))+
  geom_hline(yintercept = 0,col="grey") +
  theme(legend.position="none")

Cafreq1<-afreq1_N[,colnames(afreq1_N)%in%colnames(afreq2_N)]
Cafreq2<-afreq2_N[,colnames(afreq2_N)%in%colnames(afreq1_N)]

Dafreq<-Cafreq2-Cafreq1
Dafreq<-data.frame(rowMeans(Cafreq2)-rowMeans(Cafreq1))
Dafreq$Pops<-Pops
Dafreq$Pops <- factor(Dafreq$Pops, levels = Dafreq$Pops[1:10])

ggplot(data = Dafreq, aes(colour=Pops))+
  geom_boxplot(aes(y=Dafreq[,-length(Dafreq)],x=DM2$Pops))+
  xlab("")+ 
  ylab("Results of all neutral Allele Freq(future) - Allele Freq(present)")+
  scale_color_manual(values=colfunc(10))+
  geom_hline(yintercept = 0,col="grey") +
  theme(legend.position="none")


Comp_M2<-cbind(DcumImp,DM2)
colnames(Comp_CI_FreI)<-c("DCumImp","Pops","Extr",paste("D",seq(1,10,1),sep=""))
Comp_CI_FreI$Pops<-row.names(Comp_CI_FreI)
Comp_CI_FreI

mComp_CI_FreI_all<-merge(DallFreInd_all,DcumImp,by="Pops")
mComp_CI_FreI_all$Pops <- factor(mComp_CI_FreI_all$Pops, levels = unique(Pops))
mComp_CI_FreI_all