library(OutFLANK)  # outflank package
library(vcfR)
setwd("/Users/akijarl/Desktop/PostDoc/SLiMstuff/SLiM_output/250K/VCF_output/")
setwd("/Users/akijarl/Desktop/PostDoc/SLiMstuff/SLiM_output/")

vcf1 <- read.vcfR("TTT_SP_100_1599594485287_Gen250000.vcf", verbose = T)
vcf2 <- read.vcfR("TTT_SP_100_1599594485287_Gen250300.vcf", verbose = T)

vcfTSR <- read.vcfR("TTT_SP_100_linear_EnvShift_10RC_1649683388434_recap_mutate.vcf", verbose = T)
vcfTSR <- read.vcfR("TTT_SP_100_linear_EnvShift_10RC_1687770438261_recap_mutate.vcf", verbose = T)


queryMETA(vcfTSR)

#Tvcf<-vcfR2tidy(vcf) #Takes forever, don't do with large files
###############################################################
#Katie's code reference
###############################################################
#library(vcfR) 
#vcf <- read.vcfR(paste( seed, "_", type, ".vcf.gz", sep=""))

### Convert VCF to 012 format ####
# Character matrix containing the genotypes
# individuals in columns
# Remove 1st column, which is 'Format'
geno1 <- vcf1@gt[,-1] 
geno2 <- vcf2@gt[,-1] 

genoTSR <- vcfTSR@gt[,-1] 

position1 <- getPOS(vcf1) # Positions in bp
chromosome1 <- getCHROM(vcf1) # Chromosome information
ID1 <- getINFO(vcf1) # Meta data information

position2 <- getPOS(vcf2) # Positions in bp
chromosome2 <- getCHROM(vcf2) # Chromosome information
ID2 <- getINFO(vcf2) # Meta data information

positionTSR <- getPOS(vcfTSR) # Positions in bp
chromosomeTSR <- getCHROM(vcfTSR) # Chromosome information
#IDTSR <- getINFO(vcfTSR) # Meta data information

G1 <- matrix(NA, nrow = nrow(geno1), ncol = ncol(geno1))
G1[geno1 %in% c("0/0", "0|0")] <- 0
G1[geno1  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
G1[geno1 %in% c("1/1", "1|1")] <- 2

G2 <- matrix(NA, nrow = nrow(geno2), ncol = ncol(geno2))
G2[geno2 %in% c("0/0", "0|0")] <- 0
G2[geno2  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
G2[geno2 %in% c("1/1", "1|1")] <- 2

GTSR <- matrix(NA, nrow = nrow(genoTSR), ncol = ncol(genoTSR))
GTSR[genoTSR %in% c("0/0", "0|0")] <- 0
GTSR[genoTSR  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
GTSR[genoTSR %in% c("1/1", "1|1")] <- 2

sum(duplicated(position1))
sum(duplicated(position2))

sum(duplicated(positionTSR))

#Position of M2 allele
position1[grep("MT=2",ID1)]
position2[grep("MT=2",ID2)]

a_freq1 <- rowSums(G1)/(2*ncol(G1))
a_freq2 <- rowSums(G2)/(2*ncol(G2))
a_freqTSR <- rowSums(GTSR)/(2*ncol(GTSR))

hist(a_freq1, breaks=75,xlab="Allele frequency", main="Generation 250000, pre env. shift", col="cyan")
hist(a_freq2, breaks=75, xlab="Allele frequency", main="Generation 250300, post env. shift", col="orange")
hist(a_freqTSR, breaks=75, xlab="Allele frequency", main="From TSR file, Generation 250300, post env. shift", col="red")

plot(a_freq1[order(a_freq1)], ylab="Allele frequency", main="Generation 250000, pre env. shift", col="cyan")

plot(a_freq2[order(a_freq2)], ylab="Allele frequency", main="Generation 250300, post env. shift", col="orange")

plot(a_freqTSR[order(a_freqTSR)], ylab="Allele frequency", main="From TSR file, Generation 250300, post env. shift", col="red")

PreN<-data.frame(a_freq1,position1,ID1,stringsAsFactors = F)
PreN[grep("MT=2",PreN$ID1),]

library(stringr)
pat <- "MID=([^;]+);"
MID_pre<-str_match(PreN$ID1, pat)[,2]

PostN<-data.frame(a_freq2,position2,ID2)
PostN[grep("MT=2",PostN$ID2),]

MID_post<-str_match(PostN$ID2, pat)[,2]
