require(gradientForest)
#require(gdm)

envPop <- data.frame(seq(-1,1,length.out=9))
names(envPop) <- "envSelect"

alFreq<-data.frame(seq(0.1,0.9,0.1))
colnames(alFreq)<-"LP"

all_freq<-alFreq

alFreq<-data.frame(seq(0.9,0.1,-0.1))
colnames(alFreq)<-"LN"

all_freq$LN<-alFreq$LN

envPop <- data.frame(seq(-1,1,length.out=10))
names(envPop) <- "envSelect"

alFreq<-data.frame(c(seq(0.9,0.1,-0.2),seq(0.1,0.9,0.2)))
colnames(alFreq)<-"NLV"

all_freq<- alFreq
  
#all_freq$NLV<-alFreq$NLV

alFreq<-data.frame(c(seq(0.1,0.9,0.2),seq(0.9,0.1,-0.2))) 
colnames(alFreq)<-"NLA"

all_freq$NLA<-alFreq$NLA

##############################################
# Chunk to fit GF models to minor allele frequencies at the level of
# populations
# GF is fit to each SNP individually to 
# ease computational / memory burden

alFreq <- data.frame(all_freq$LP)
colnames(alFreq)<-"LP"

gfMod <- gradientForest(data=data.frame(envPop, alFreq),
                        predictor.vars=colnames(envPop),
                        response.vars=colnames(alFreq),
                        corr.threshold=0.5, 
                        ntree=500, 
                        trace=T)

# Calculate genomic offset
# note that I am doing this for the avearge across all alleles since 
# GF was fit to all alleles simultaneously
# The more correct way is to calculate offset for adaptive alleles only,
# either individually or for a model fit to just those alleles.
gfTrans1 <- predict(gfMod, envPop)
colnames(gfTrans1)<-"C.Imp_genome_before"

#gfTrans2 <- predict(gfMod, envPop.shift)
#colnames(gfTrans2)<-"C.Imp_genome_after"

# offset needs to be considered using absolute values ()
#offset <- gfTrans2-gfTrans1
#offset <- gfTrans2-gfTrans1
#colnames(offset)<-"D_C.Imp_genome"


#Visual of linear CI
cols <- c("AF increasing w/ env"="#0000FFBF","AF decreasing w/ env"="#FF0000BF")
ggplot() + 
  geom_line(aes(x=envPop$envSelect, y=gfTrans1$C.Imp_genome_before, colour="AF increasing w/ env"), lwd=1) + #blue
  labs(y="GF Cumulative Importance", x="Environment") +
  geom_line(aes(x=envPop$envSelect, y=gfTrans1$C.Imp_genome_before,colour="AF decreasing w/ env"), lwd=1) + #red
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  scale_colour_manual(name="",values=cols) + 
  ggtitle("Linear relationship between allele frequency (AF) and environment")+
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  theme(plot.title = element_text(size=14, face="bold.italic"))

#Visual of non-linear CI
cols <- c("Stabilizing"="#0000FFBF","Disruptive"="#FF0000BF")
ggplot() + 
  geom_line(aes(x=envPop$envSelect, y=gfTrans1_N$C.Imp_genome_before, colour="Stabilizing"), lwd=1) + #blue
  labs(y="GF Cumulative Importance", x="Environment") +
  geom_line(aes(x=envPop$envSelect, y=gfTrans1_U$C.Imp_genome_before,colour="Disruptive"), lwd=1) + #red
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  scale_colour_manual(name="",values=cols) + 
  ggtitle("Non-linear relationship between allele frequency (AF) and environment")+
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  theme(plot.title = element_text(size=14, face="bold.italic"))


plot(envPop$envSelect,all_freq$LN,type="l",col="blue", xlab="Environment", ylab="Allele frequency",lwd=2.5, yaxp  = c(0.1, 0.9, 8), las=1)
lines(envPop$envSelect,all_freq$LP,col="red",lwd=2.5)

plot(envPop$envSelect,all_freq$NLV,type="l",col="blue", xlab="Environment", ylab="Allele frequency",lwd=2.5, yaxp  = c(0.1, 0.9, 8), las=1)
lines(envPop$envSelect,all_freq$NLA,col="red",lwd=2.5)
