#Visualizing multilocus data
require(ggplot2)
require(cowplot)
require(qqman)
require(RColorBrewer)

colfunc<-colorRampPalette(c("orange","cyan"))

colour1 = rep(colfunc(10),10)

par(mfrow=c(2,1))
par(mar = c(1,4,4,0))
plot(fit$Gen,fit$P19_phen1,type="l", main=plotTitle, xlab="",xaxt="n", ylab="Phenotype 1",col="white",ylim=c(-4,4))
k=1
for(i in which(fitt$Type=="Phen1")){
  lines(fit$Gen,fit[,i+13],lwd=2,col=colour1[k])
  k=k+1
}
par(mar = c(5,4,0,0))
colour2 = rep(colfunc(10), each=10)
plot(fit$Gen,fit$P19_phen1,type="l", main=NULL, xlab="Generations",ylab="Phenotype 2",col="white",ylim=c(-4,4))
k=1
for(i in which(fitt$Type=="Phen2")){
  lines(fit$Gen,fit[,i+13],lwd=2,col=colour2[k])
  k=k+1
}

dev.off()

plot(fit$Gen,fit$P19_env1,type="l", main=plotTitle, xlab="Generations",ylab="Environment 1",col="white",ylim=c(-1,1))
k=1
for(i in which(fitt$Type=="Env1")){
  lines(fit$Generation,fit[,i+13],lwd=2,col=colour1[k])
  k=k+1
}

plot(fit$Generation,fit$P1_env2,type="l", main=plotTitle, xlab="Generations",ylab="Environment 2",col="white",ylim=c(-1,1))
color=rep(colfunc(10), each=10)
k=1
for(i in which(fitt$Type=="Env2")){
  lines(fit$Generation,fit[,(i+13)],lwd=2,col=colour2[k])
  k=k+1
}

plot(fit$Gen,fit$P19_fit,type="l",main=plotTitle ,xlab="Generations",ylab="Absolute Fitness",col="red", ylim=c(0.6,1))
x=1:10
y=1:10
for(i in x){
  for(j in y){
    k=((i - 1) + (j - 1) * 10 + 13)
    #print(paste("k = ", k))
    lines(fit$Gen,fit[,k],col=colour1[i])
  }
}

#GF visualization per environment
A <- ggplot() +
  geom_line(aes(x=envPop1$envSelect, y=gfTrans1e1$C.Imp_genome_before), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  ylab("Cumulative Importance\n(Genome)")+
  xlab("") +
  #theme(axis.title.y = element_text(margin = margin(t = 50, r = 0, b = 0, l = 50)))+
  #labs(y="Cumulative Importance\n(Genome)", x="Environment 1") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle(plotTitle) +
    #ggtitle(paste(plotTitle,", all alleles")) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

B <- ggplot() +
  geom_line(aes(x=envPop2$envSelect, y=gfTrans1e2$C.Imp_genome_before), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="Cumulative Importance\n(Genome)\n", x="") +
  #labs(y="Cumulative Importance\n(Genome)", x="Environment 2") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  #ggtitle(plotTitle) +
  scale_y_continuous(position = "right", sec.axis = sec_axis(~., labels = NULL)) +
  theme(plot.title = element_text(size=14, face="bold.italic"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

#GF visualization per environment with the just alleles under selection
C <- ggplot() +
  geom_line(aes(x=envPop1$envSelect, y=gfTrans1_sel1$C.Imp_genome_before), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="\n(Adaptive)", x="Environment 1") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  #ggtitle(paste(plotTitle,", alleles under selection")) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))

D <- ggplot() +
  geom_line(aes(x=envPop2$envSelect, y=gfTrans1_sel2$C.Imp_genome_before), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  ylab("\n(Adaptive)\n") +
  xlab("Environment 2")+
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  #theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  scale_y_continuous(position = "right", sec.axis = sec_axis(~., labels = NULL)) +
  #ggtitle(plotTitle) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))

title <- ggdraw() + draw_label(plotTitle, fontface='bold')
plot_grid(title,NULL,A,B,C,D,nrow=3,ncol=2,rel_heights = c(1,3,3))


##################################################################
# Common garden visualizations
##################################################################
# SIGMA_K = 0.5
# SIGMA_STAT = 4
# BURNIN_1 = 10000
# BURNIN_2 = 12000
# generation = 10000:12000
#
# sigmak <- NULL
# for(i in 1:length(generation)){
#   sigmak <- c(sigmak,(SIGMA_K * ((generation[i]- BURNIN_1) / (BURNIN_2 - BURNIN_1)) + SIGMA_STAT * (1.0 - (generation[i]- BURNIN_1) / (BURNIN_2 - BURNIN_1))))
# }
# plot(generation,sigmak)

cg_sum<-read.table(paste("results/SLiM_output/CG_files/",seed,"_ML_WF_CG_sum_Gen.txt",sep=""),header=T)

ggplot(data=cg_sum)+
  #geom_line(aes(y=sympatry, x=generation), col="green")+
  #geom_line(aes(y=allopatry, x=generation), col="orange")+
  geom_line(aes(y=local_adaptation, x=generation), col="cyan", lwd=2)+
  geom_vline(xintercept = 1000, linetype='dotted')+
  geom_vline(xintercept = 2000, linetype='dotdash')+
  geom_vline(xintercept = 3000, linetype='dashed')+
  xlab("Generations")+
  ylab("Local Adaptation")+
  annotate("text", y=-0.01, x=500, label= "Burnin", size=2.75)+
  annotate("text", y=-0.01, x=1500, label= "Env.\nTrans.", size=2.75)+
  annotate("text", y=-0.01, x=2200, label= "Adaptation", size=2.75)+
  annotate("text", y=-0.01, x=3200, label= "Climate\nChange", size=2.75)+
  ggtitle(paste(plotTitle)) +
  theme_classic()

ggplot(data=cg_sum)+
  #geom_line(aes(y=sympatry, x=generation), col="green")+
  #geom_line(aes(y=allopatry, x=generation), col="orange")+
  geom_line(aes(y=corr_pheno0_opt0, x=generation), col="orange", lwd=2)+
  geom_line(aes(y=corr_pheno1_opt1, x=generation), col="cyan", lwd=2)+
  geom_vline(xintercept = 1000, linetype='dotted')+
  geom_vline(xintercept = 2000, linetype='dotdash')+
  geom_vline(xintercept = 3000, linetype='dashed')+
  ylim(-1,1)+
  xlab("Generations")+
  ylab("Correlation of phenotypes to optima")+
  annotate("text", y=-1, x=500, label= "Burnin", size=2.75)+
  annotate("text", y=-1, x=1500, label= "Env.\nTrans.", size=2.75)+
  annotate("text", y=-1, x=2200, label= "Adaptation", size=2.75)+
  annotate("text", y=-1, x=3200, label= "Climate\nChange", size=2.75)+
  ggtitle(paste(plotTitle)) +
  theme_classic()

cg<-read.table(paste("results/SLiM_output/CG_files/",seed,"_fitnessmat_pop.txt",sep=""),header=F)

colnames(cg)<-rep(paste("H",seq(1:100),sep=""))
cg$Transplant<-paste("T",seq(1:100),sep="")

cg_df <- melt(cg,  id.vars ="Transplant",  variable.name = "Home",
              value.name = "Fitness")

cg_df$Transplant<-factor(cg_df$Transplant,levels=unique(cg_df$Transplant))

CI_bf1 <- gfTrans1e1$C.Imp_genome_before
CI_bf2 <- gfTrans1e2$C.Imp_genome_before

GF_dist_bef<-data.frame(cbind(gfTrans1e1,gfTrans1e2))
colnames(GF_dist_bef)<-c("CI_env1", "CI_env2")

ggplot() +
  geom_line(aes(x=1:100, y=GF_dist_bef$CI_env1) ,col="orange", lty=2,size=1.5)+
  geom_line(aes(x=1:100, y=GF_dist_bef$CI_env2) ,col="cyan", lty=2,size=1.5)+
  xlab("Location")+
  ylab("Cumulative Importance")


CI<-NULL
for(i in 1:nrow(GF_dist_bef)){
  for(j in 1:nrow(GF_dist_bef)){
    CI<-c(CI,dist(rbind(GF_dist_bef[i,],GF_dist_bef[j,])))
    #CI<-c(CI,CI_bf[j]-CI_bf[i])
  }
}

cg_df$D_CI<-CI


GFsel_dist_bef<-data.frame(cbind(gfTrans1_sel1,gfTrans1_sel2))
colnames(GFsel_dist_bef)<-c("CI_env1", "CI_env2")

CI_sel<-NULL
for(i in 1:nrow(GFsel_dist_bef)){
  for(j in 1:nrow(GFsel_dist_bef)){
    CI_sel<-c(CI_sel,dist(rbind(GFsel_dist_bef[i,],GFsel_dist_bef[j,])))
    #CI<-c(CI,CI_bf[j]-CI_bf[i])
  }
}

cg_df$D_CI_sel<-CI_sel

cg_df$FST<-as.vector(Pre_FST)

cg_df$FST_sel<-as.vector(Pre_FST_sel)




demes <-unique(cg_df$Transplant)
FST_plots <- vector('list', length(demes))
max(cg_df[!is.na(cg_df$FST),]$FST)
min(cg_df[!is.na(cg_df$FST),]$FST)
for (i in seq_along(demes)) {
  dat_filt<-cg_df[cg_df$Transplant%in%demes[i],]
  dat_sum<-cor.test(x=dat_filt$FST, y=dat_filt$Fitness, method = "pearson")
  FST_plots[[i]] <- local({
    i <- i
    p1 <- ggplot(data=dat_filt)+
      geom_smooth(aes(x=FST, y=Fitness), method="lm",se = F)+
      annotate("text",x=0.002,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
      geom_point(aes(x=FST, y=Fitness, col=Home), size=3)+
      labs(shape="Home", colour="Common\nGarden\nLocation")+
      scale_color_manual(values = colfunc(100), guide=F)+
      # ylab("Average relative fitness in common garden")+
      # xlab("Pairwise FST (genome)")+
      ylab("")+
      xlab("")+
      ylim(0,1)+
      xlim(-0.003,0.008)+
      ggtitle(paste("Transplant to",str_sub(demes[i],2,4)))+
      theme_classic()+
      theme(axis.text=element_text(size=12),axis.title=element_text(size=14)
  )
    print(p1)
  })
}

plot_grid(FST_plots[[1]],FST_plots[[2]],FST_plots[[3]],FST_plots[[4]],FST_plots[[5]],FST_plots[[6]],FST_plots[[7]],FST_plots[[8]],FST_plots[[9]],FST_plots[[10]],
          FST_plots[[11]],FST_plots[[12]],FST_plots[[13]],FST_plots[[14]],FST_plots[[15]],FST_plots[[16]],FST_plots[[17]],FST_plots[[18]],FST_plots[[19]],FST_plots[[20]],
          FST_plots[[21]],FST_plots[[22]],FST_plots[[23]],FST_plots[[24]],FST_plots[[25]],FST_plots[[26]],FST_plots[[27]],FST_plots[[28]],FST_plots[[29]],FST_plots[[30]],
          FST_plots[[31]],FST_plots[[32]],FST_plots[[33]],FST_plots[[34]],FST_plots[[35]],FST_plots[[36]],FST_plots[[37]],FST_plots[[38]],FST_plots[[39]],FST_plots[[40]],
          FST_plots[[41]],FST_plots[[42]],FST_plots[[43]],FST_plots[[44]],FST_plots[[45]],FST_plots[[46]],FST_plots[[47]],FST_plots[[48]],FST_plots[[49]],FST_plots[[50]],
          nrow=5)

plot_grid(FST_plots[[51]],FST_plots[[52]],FST_plots[[53]],FST_plots[[54]],FST_plots[[55]],FST_plots[[56]],FST_plots[[57]],FST_plots[[58]],FST_plots[[59]],FST_plots[[60]],
          FST_plots[[66]],FST_plots[[62]],FST_plots[[63]],FST_plots[[64]],FST_plots[[65]],FST_plots[[66]],FST_plots[[67]],FST_plots[[68]],FST_plots[[69]],FST_plots[[70]],
          FST_plots[[71]],FST_plots[[77]],FST_plots[[73]],FST_plots[[74]],FST_plots[[75]],FST_plots[[76]],FST_plots[[77]],FST_plots[[78]],FST_plots[[79]],FST_plots[[80]],
          FST_plots[[81]],FST_plots[[82]],FST_plots[[88]],FST_plots[[84]],FST_plots[[85]],FST_plots[[86]],FST_plots[[87]],FST_plots[[88]],FST_plots[[89]],FST_plots[[90]],
          FST_plots[[91]],FST_plots[[92]],FST_plots[[93]],FST_plots[[94]],FST_plots[[95]],FST_plots[[96]],FST_plots[[97]],FST_plots[[98]],FST_plots[[99]],FST_plots[[100]],
          nrow=5)

FST_ad_plots <- vector('list', length(demes))
max(cg_df[!is.na(cg_df$FST_sel),]$FST_sel)
min(cg_df[!is.na(cg_df$FST_sel),]$FST_sel)
for (i in seq_along(demes)) {
  dat_filt<-cg_df[cg_df$Transplant%in%demes[i],]
  dat_sum<-cor.test(x=dat_filt$FST_sel, y=dat_filt$Fitness, method = "pearson")
  FST_ad_plots[[i]] <- local({
    i <- i
    p1 <- ggplot(data=dat_filt)+
      geom_smooth(aes(x=FST_sel, y=Fitness), method="lm",se = F)+
      annotate("text",x=0.006,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
      geom_point(aes(x=FST_sel, y=Fitness, col=Home), size=3)+
      labs(shape="Home", colour="Common\nGarden\nLocation")+
      scale_color_manual(values = colfunc(100), guide=F)+
      # ylab("Average relative fitness in common garden")+
      # xlab("Pairwise FST (adapted)")+
      ylab("")+
      xlab("")+
      ylim(0,1)+
      xlim(-0.03,0.05)+
      ggtitle(paste("Transplant to",str_sub(demes[i],2,4)))+
      theme_classic()+
      theme(axis.text=element_text(size=12),axis.title=element_text(size=14)
      )
    print(p1)
  })
}

plot_grid(FST_ad_plots[[1]],FST_ad_plots[[2]],FST_ad_plots[[3]],FST_ad_plots[[4]],FST_ad_plots[[5]],FST_ad_plots[[6]],FST_ad_plots[[7]],FST_ad_plots[[8]],FST_ad_plots[[9]],FST_ad_plots[[10]],
          FST_ad_plots[[11]],FST_ad_plots[[12]],FST_ad_plots[[13]],FST_ad_plots[[14]],FST_ad_plots[[15]],FST_ad_plots[[16]],FST_ad_plots[[17]],FST_ad_plots[[18]],FST_ad_plots[[19]],FST_ad_plots[[20]],
          FST_ad_plots[[21]],FST_ad_plots[[22]],FST_ad_plots[[23]],FST_ad_plots[[24]],FST_ad_plots[[25]],FST_ad_plots[[26]],FST_ad_plots[[27]],FST_ad_plots[[28]],FST_ad_plots[[29]],FST_ad_plots[[30]],
          FST_ad_plots[[31]],FST_ad_plots[[32]],FST_ad_plots[[33]],FST_ad_plots[[34]],FST_ad_plots[[35]],FST_ad_plots[[36]],FST_ad_plots[[37]],FST_ad_plots[[38]],FST_ad_plots[[39]],FST_ad_plots[[40]],
          FST_ad_plots[[41]],FST_ad_plots[[42]],FST_ad_plots[[43]],FST_ad_plots[[44]],FST_ad_plots[[45]],FST_ad_plots[[46]],FST_ad_plots[[47]],FST_ad_plots[[48]],FST_ad_plots[[49]],FST_ad_plots[[50]],
          nrow=5)

plot_grid(FST_ad_plots[[51]],FST_ad_plots[[52]],FST_ad_plots[[53]],FST_ad_plots[[54]],FST_ad_plots[[55]],FST_ad_plots[[56]],FST_ad_plots[[57]],FST_ad_plots[[58]],FST_ad_plots[[59]],FST_ad_plots[[60]],
          FST_ad_plots[[66]],FST_ad_plots[[62]],FST_ad_plots[[63]],FST_ad_plots[[64]],FST_ad_plots[[65]],FST_ad_plots[[66]],FST_ad_plots[[67]],FST_ad_plots[[68]],FST_ad_plots[[69]],FST_ad_plots[[70]],
          FST_ad_plots[[71]],FST_ad_plots[[77]],FST_ad_plots[[73]],FST_ad_plots[[74]],FST_ad_plots[[75]],FST_ad_plots[[76]],FST_ad_plots[[77]],FST_ad_plots[[78]],FST_ad_plots[[79]],FST_ad_plots[[80]],
          FST_ad_plots[[81]],FST_ad_plots[[82]],FST_ad_plots[[88]],FST_ad_plots[[84]],FST_ad_plots[[85]],FST_ad_plots[[86]],FST_ad_plots[[87]],FST_ad_plots[[88]],FST_ad_plots[[89]],FST_ad_plots[[90]],
          FST_ad_plots[[91]],FST_ad_plots[[92]],FST_ad_plots[[93]],FST_ad_plots[[94]],FST_ad_plots[[95]],FST_ad_plots[[96]],FST_ad_plots[[97]],FST_ad_plots[[98]],FST_ad_plots[[99]],FST_ad_plots[[100]],
          nrow=5)

CI_plots <- vector('list', length(demes))
max(cg_df[!is.na(cg_df$FST_sel),]$FST_sel)
min(cg_df[!is.na(cg_df$FST_sel),]$FST_sel)
for (i in seq_along(demes)) {
  dat_filt<-cg_df[cg_df$Transplant%in%demes[i],]
  dat_sum<-cor.test(x=dat_filt$D_CI, y=dat_filt$Fitness, method = "pearson")
  CI_plots[[i]] <- local({
    i <- i
    p1 <- ggplot(data=dat_filt)+
      geom_smooth(aes(x=D_CI, y=Fitness), method="lm",se = F)+
      annotate("text",x=0.025,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
      geom_point(aes(x=D_CI, y=Fitness, col=Home), size=3)+
      labs(shape="Home", colour="Common\nGarden\nLocation")+
      scale_color_manual(values = colfunc(100), guide=F)+
      ylab("Average relative fitness in common garden")+
      xlab("GF Offset (genome)")+
      ylim(0,1)+
      xlim(0,0.06)+
      ggtitle(paste("Transplant to",str_sub(demes[i],2,4)))+
      theme_classic()+
      theme(axis.text=element_text(size=12),axis.title=element_text(size=14)
      )
    print(p1)
  })
}

CI_ad_plots <- vector('list', length(demes))

for (i in seq_along(demes)) {
  dat_filt<-cg_df[cg_df$Transplant%in%demes[i],]
  dat_sum<-cor.test(x=dat_filt$D_CI_sel, y=dat_filt$Fitness, method = "pearson")
  CI_ad_plots[[i]] <- local({
    i <- i
    p1 <- ggplot(data=dat_filt)+
      geom_smooth(aes(x=D_CI_sel, y=Fitness), method="lm",se = F)+
      annotate("text",x=0.025,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
      geom_point(aes(x=D_CI_sel, y=Fitness, col=Home), size=3)+
      labs(shape="Home", colour="Common\nGarden\nLocation")+
      scale_color_manual(values = colfunc(100), guide=F)+
      ylab("Average relative fitness in common garden")+
      xlab("GF Offset (adapted)")+
      ylim(0,1)+
      xlim(0,0.13)+
      ggtitle(paste("Transplant to",str_sub(demes[i],2,4)))+
      theme_classic()+
      theme(axis.text=element_text(size=12),axis.title=element_text(size=14)
      )
    print(p1)
  })
}


 #demes <-c("T34","T35","T36","T37","T44","T45","T46","T47","T54","T55","T56","T57")
demes <-c("T34","T35","T36","T37","T44","T45","T46","T47","T54","T55","T56","T57")
for (i in 1:length(demes)){
  dat_filt<-cg_df[cg_df$Transplant%in%demes[i],]
  dat_sum<-cor.test(x=dat_filt$D_CI, y=dat_filt$Fitness, method = "pearson")

  print(ggplot(data=dat_filt)+
          geom_smooth(aes(y=Fitness, x=D_CI), method="lm",se = F)+
          annotate("text",x=0,y=-04.,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
          geom_point(aes(x=D_CI, y=Fitness, col=Home), size=3)+
          labs(shape="Home", colour="Common\nGarden\nLocation")+
          scale_color_manual(values = colfunc(100), guide=F)+
          ylab("Average relative fitness in common garden")+
          xlab("Absolute GF Offset (genome)")+
          #ylim(0,1)+
          #xlim(0,0.3)+
          ggtitle(paste("Transplant to",str_sub(demes[i],2,3)))+
          theme_classic()+
          theme(axis.text=element_text(size=12),axis.title=element_text(size=14))
  )
}

dat_sum<-cor.test(y=dat_filt$Fitness,x=dat_filt$D_CI_sel, method = "pearson")

ggplot(data=dat_filt)+
  geom_smooth(aes(y=Fitness, x=D_CI_sel), method="lm",se = F)+
  annotate("text",x=0.05,y=0.25,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
  geom_point(aes(x=D_CI_sel, y=Fitness, col=Home), size=3)+
  labs(shape="Home", colour="Common\nGarden\nLocation")+
  scale_color_manual(values = colfunc(100), guide=F)+
  ylab("Average relative fitness in common garden")+
  xlab("Absolute GF Offset (allele)")+
  ylim(0,1)+
  xlim(0,0.3)+
  ggtitle("Transplant to T34")+
  theme_classic()+
  theme(axis.text=element_text(size=12),axis.title=element_text(size=14))

#demes <-c("T1","T2","T11","T12","T21","T22","T9","T10","T19","T20","T29","T30")
demes <-c("T34","T35","T36","T37","T44","T45","T46","T47","T54","T55","T56","T57")
for (i in demes){
  dat_filt<-cg_df[cg_df$Transplant%in%c(i),]
  dat_sum<-cor.test(x=dat_filt$FST, y=dat_filt$Fitness, method = "pearson")

  print(ggplot(data=dat_filt)+
          geom_smooth(aes(x=FST, y=Fitness), method="lm",se = F)+
          annotate("text",x=0.006,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
          geom_point(aes(x=FST, y=Fitness, col=Home), size=3)+
          labs(shape="Home", colour="Common\nGarden\nLocation")+
          scale_color_manual(values = colfunc(100), guide=F)+
          ylab("Average relative fitness in common garden")+
          xlab("Pairwise FST (genome)")+
          ylim(0,1)+
          xlim(0,0.007)+
          ggtitle(paste("Transplant to",str_sub(i,2,3)))+
          theme_classic()+
          theme(axis.text=element_text(size=12),axis.title=element_text(size=14))
  )
}

#demes <-c("T1","T2","T11","T12","T21","T22","T9","T10","T19","T20","T29","T30")
demes <-c("T34","T35","T36","T37","T44","T45","T46","T47","T54","T55","T56","T57")

for (i in demes){
  dat_filt<-cg_df[cg_df$Transplant%in%c(i),]
  dat_sum<-cor.test(x=dat_filt$FST_sel, y=dat_filt$Fitness, method = "pearson")

  print(ggplot(data=dat_filt)+
          geom_smooth(aes(x=FST_sel, y=Fitness), method="lm",se = F)+
          annotate("text",x=0.03,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
          geom_point(aes(x=FST_sel, y=Fitness, col=Home), size=3)+
          labs(shape="Home", colour="Common\nGarden\nLocation")+
          scale_color_manual(values = colfunc(100), guide=F)+
          ylab("Average relative fitness in common garden")+
          xlab("Pairwise FST (allele)")+
          ylim(0,1)+
          xlim(0,0.04)+
          ggtitle(paste("Transplant to",str_sub(i,2,3)))+
          theme_classic()+
          theme(axis.text=element_text(size=12),axis.title=element_text(size=14))
  )
}


demes <-c("T34","T35","T36","T37","T44","T45","T46","T47","T54","T55","T56","T57")

for (i in demes){
  dat_filt<-cg_df[cg_df$Transplant%in%c(i),]
  dat_sum<-cor.test(x=dat_filt$FST_sel, y=dat_filt$D_CI_sel, method = "pearson")

  print(ggplot(data=dat_filt)+
          geom_smooth(aes(x=FST_sel, y=D_CI_sel), method="lm",se = F)+
          annotate("text",x=0.03,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
          geom_point(aes(x=FST_sel, y=D_CI_sel, col=Home), size=3)+
          labs(shape="Home", colour="Common\nGarden\nLocation")+
          scale_color_manual(values = colfunc(100), guide=F)+
          ylab("Absolute GF Offset (allele)")+
          xlab("Pairwise FST (allele)")+
          ylim(0,1)+
          xlim(0,0.04)+
          ggtitle(paste("Transplant to",str_sub(i,2,3)))+
          theme_classic()+
          theme(axis.text=element_text(size=12),axis.title=element_text(size=14))
  )
}

cg_ind<-read.table(paste("results/SLiM_output/CG_files/",seed,"_fitnessmat_ind.txt",sep=""),header=T)

colnames(cg_ind)<-rep(paste("ID",seq(1:10000),sep=""))
cg_ind$Transplant<-paste("T",seq(1:100),sep="")

cg_df_ind <- melt(cg_ind,  id.vars ="Transplant",  variable.name = "Individual",
                  value.name = "Fitness")

cg_df_ind$Transplant<-factor(cg_df_ind$Transplant,levels=unique(cg_df_ind$Transplant))

Home<-NULL
for(i in 1:100){
  Home<-c(Home,rep(paste("H",i,sep=""),10000))
}
cg_df_ind$Home <- Home
cg_df_ind$Home<-factor(cg_df_ind$Home,levels=unique(cg_df_ind$Home))


CI<-NULL
for(i in 1:nrow(GF_dist_bef)){
  for(j in 1:nrow(GF_dist_bef)){
    CI<-c(CI,dist(rbind(GF_dist_bef[i,],GF_dist_bef[j,])))
    #CI<-c(CI,CI_bf[j]-CI_bf[i])
  }
}

CI_ind_rep<-NULL
for (i in seq(1,10000,100)){
  CI_ind_rep<-c(CI_ind_rep,rep(CI[i:(i+99)],100))
}

cg_df_ind$D_CI<-CI_ind_rep

GFsel_dist_bef<-data.frame(cbind(gfTrans1_sel1,gfTrans1_sel2))
colnames(GFsel_dist_bef)<-c("CI_env1", "CI_env2")

CI_sel<-NULL
for(i in 1:nrow(GFsel_dist_bef)){
  for(j in 1:nrow(GFsel_dist_bef)){
    CI_sel<-c(CI_sel,dist(rbind(GFsel_dist_bef[i,],GFsel_dist_bef[j,])))
    #CI<-c(CI,CI_bf[j]-CI_bf[i])
  }
}

CI_sel_ind_rep<-NULL
for (i in seq(1,10000,100)){
  CI_sel_ind_rep<-c(CI_sel_ind_rep,rep(CI_sel[i:(i+99)],100))
}

cg_df_ind$D_CI_sel<-CI_sel_ind_rep



dat_filt_ind<-cg_df_ind[cg_df_ind$Transplant%in%c("T1"),]

ggplot(data=dat_filt_ind)+
  geom_point(aes(x=D_CI, y=Fitness, col=Home), size=3)+
  geom_smooth(aes(y=Fitness, x=D_CI), method="lm",se = F)+
  labs(colour="Home Location")+
  scale_color_manual(values = colfunc(100))+
  #scale_shape_manual(values=c(15:115))+
  ylab("Average relative fitness in common garden")+
  xlab("Absolute GF Offset (genome)")+
  ylim(0,1)+
  xlim(0,0.25)+
  ggtitle("Transplant to T1")+
  theme_classic()+
  theme(axis.text=element_text(size=12),axis.title=element_text(size=14))




dat_filt<-cg_df_ind[cg_df_ind$Transplant%in%c("T11"),]
dat_sum<-cor.test(y=dat_filt$Fitness,x=dat_filt$D_CI_sel, method = "pearson")

ggplot(data=dat_filt)+
  annotate("text",x=0.35,y=0.75,label=paste("\u03C1 = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
  geom_point(aes(x=D_CI_sel, y=Fitness, col=Home), size=3)+
  geom_smooth(aes(y=Fitness, x=D_CI_sel), method="lm",se = F)+
  labs(shape="Home", colour="Source Location")+
  scale_color_manual(values = colfunc(100))+
  ylab("Individual relative fitness in common garden")+
  xlab("Absolute GF Offset (alleles)")+
  ylim(0,1)+
  xlim(0,0.45)+
  ggtitle("Transplant to T11")+
  theme_classic()+
  theme(axis.text=element_text(size=12),axis.title=element_text(size=14))

demes <-c("T34","T35","T36","T37","T44","T45","T46","T47","T54","T55","T56","T57")
for (i in demes){
  dat_filt<-cg_df_ind[cg_df_ind$Transplant%in%c(i),]
  dat_sum<-cor.test(y=dat_filt$Fitness,x=dat_filt$D_CI_sel, method = "pearson")

  print(ggplot(data=dat_filt)+
          annotate("text",x=0.35,y=0.75,label=paste("r =",round(dat_sum$estimate[[1]],3),"\np-value =",signif(dat_sum$p.value,3)))+
          geom_point(aes(x=D_CI_sel, y=Fitness, col=Home), alpha = 0.2, size=3)+
          geom_smooth(aes(y=Fitness, x=D_CI_sel), method="lm",se = F)+
          labs(shape="Home", colour="Source Location")+
          scale_color_manual(values = colfunc(100), guide=F)+
          ylab("Individual relative fitness in common garden")+
          xlab("Absolute GF Offset (alleles)")+
          ylim(0,1)+
          xlim(0,0.45)+
          ggtitle(paste("Transplant to",i))+
          theme_classic()+
          theme(axis.text=element_text(size=12),axis.title=element_text(size=14))
  )
}

#Visualizing multilocus data
require(ggplot2)
require(cowplot)
require(qqman)
require(RColorBrewer)

colfunc<-colorRampPalette(c("orange","cyan"))

colour1 = rep(colfunc(10),10)
plot(fit$Gen,fit$P19_phen1,type="l", main=plotTitle, xlab="Generations",ylab="Phenotype 1",col="white",ylim=c(-6,6))
k=1
for(i in which(fitt$Type=="Phen1")){
  lines(fit$Gen,fit[,i+13],lwd=2,col=colour1[k])
  k=k+1
}

colour2 = rep(colfunc(10), each=10)
plot(fit$Gen,fit$P19_phen1,type="l", main=plotTitle, xlab="Generations",ylab="Phenotype 2",col="white",ylim=c(-6,6))
k=1
for(i in which(fitt$Type=="Phen2")){
  lines(fit$Gen,fit[,i+13],lwd=2,col=colour2[k])
  k=k+1
}

plot(fit$Gen,fit$P19_env1,type="l", main=plotTitle, xlab="Generations",ylab="Environment 1",col="white",ylim=c(-1,1))
k=1
for(i in which(fitt$Type=="Env1")){
  lines(fit$Generation,fit[,i+13],lwd=2,col=colour1[k])
  k=k+1
}

plot(fit$Generation,fit$P1_env2,type="l", main=plotTitle, xlab="Generations",ylab="Environment 2",col="white",ylim=c(-1,1))
color=rep(colfunc(10), each=10)
k=1
for(i in which(fitt$Type=="Env2")){
  lines(fit$Generation,fit[,(i+13)],lwd=2,col=colour2[k])
  k=k+1
}

plot(fit$Gen,fit$P19_fit,type="l",main=plotTitle ,xlab="Generations",ylab="Absolute Fitness",col="red", ylim=c(0.7,1))
x=1:10
y=1:10
for(i in x){
  for(j in y){
    k=((i - 1) + (j - 1) * 10 + 13)
    #print(paste("k = ", k))
    lines(fit$Gen,fit[,k],col=colour1[i])
  }
}

#GF visualization per environment
ggplot() +
  geom_line(aes(x=envPop1$envSelect, y=gfTrans1e1$C.Imp_genome_before), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="Cumulative Importance", x="Environment 1") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  ggtitle(paste(plotTitle,", all alleles")) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))

#GF visualization per environment with the just alleles under selection
ggplot() +
  geom_line(aes(x=envPop1$envSelect, y=gfTrans1_sel1$C.Imp_genome_before), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="Cumulative Importance", x="Environment 1") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  ggtitle(paste(plotTitle,", alleles under selection")) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))


##################################################################
# Common garden visualizations
##################################################################
SIGMA_K = 0.5
SIGMA_STAT = 4
BURNIN_1 = 10000
BURNIN_2 = 12000
generation = 10000:12000

sigmak <- NULL
for(i in 1:length(generation)){
  sigmak <- c(sigmak,(SIGMA_K * ((generation[i]- BURNIN_1) / (BURNIN_2 - BURNIN_1)) + SIGMA_STAT * (1.0 - (generation[i]- BURNIN_1) / (BURNIN_2 - BURNIN_1))))
}
plot(generation,sigmak)

cg_sum<-read.table(paste("results/SLiM_output/CG_files/",seed,"_ML_WF_CG_sum_Gen.txt",sep=""),header=T)

ggplot(data=cg_sum)+
  #geom_line(aes(y=sympatry, x=generation), col="green")+
  #geom_line(aes(y=allopatry, x=generation), col="orange")+
  geom_line(aes(y=local_adaptation, x=generation), col="cyan")+
  geom_vline(xintercept = 1000, linetype='dotted')+
  geom_vline(xintercept = 2000, linetype='dotdash')+
  geom_vline(xintercept = 3000, linetype='dashed')+
  xlab("Generations")+
  ylab("Local Adaptation")+
  annotate("text", y=-0.01, x=500, label= "Burnin", size=2.75)+
  annotate("text", y=-0.01, x=1500, label= "Env.\nTrans.", size=2.75)+
  annotate("text", y=-0.01, x=2200, label= "Adaptation", size=2.75)+
  annotate("text", y=-0.01, x=3200, label= "Climate\nChange", size=2.75)+
  ggtitle(paste(plotTitle)) +
  theme_classic()

 ggplot(data=cg_sum)+
  #geom_line(aes(y=sympatry, x=generation), col="green")+
  #geom_line(aes(y=allopatry, x=generation), col="orange")+
  geom_line(aes(y=corr_pheno0_opt0, x=generation), col="orange")+
  geom_line(aes(y=corr_pheno1_opt1, x=generation), col="cyan")+
  geom_vline(xintercept = 1000, linetype='dotted')+
  geom_vline(xintercept = 2000, linetype='dotdash')+
  geom_vline(xintercept = 3000, linetype='dashed')+
  ylim(-1,1)+
  xlab("Generations")+
  ylab("Correlation of phenotypes to optima")+
  annotate("text", y=-1, x=500, label= "Burnin", size=2.75)+
  annotate("text", y=-1, x=1500, label= "Env.\nTrans.", size=2.75)+
  annotate("text", y=-1, x=2200, label= "Adaptation", size=2.75)+
  annotate("text", y=-1, x=3200, label= "Climate\nChange", size=2.75)+
  ggtitle(paste(plotTitle)) +
  theme_classic()

cg<-read.table(paste("results/SLiM_output/CG_files/",seed,"_fitnessmat_pop.txt",sep=""),header=F)

colnames(cg)<-rep(paste("H",seq(1:100),sep=""))
cg$Transplant<-paste("T",seq(1:100),sep="")

cg_df <- melt(cg,  id.vars ="Transplant",  variable.name = "Home",
              value.name = "Fitness")

cg_df$Transplant<-factor(cg_df$Transplant,levels=unique(cg_df$Transplant))

CI_bf1 <- gfTrans1e1$C.Imp_genome_before
CI_bf2 <- gfTrans1e2$C.Imp_genome_before

GF_dist_bef<-data.frame(cbind(gfTrans1e1,gfTrans1e2))
colnames(GF_dist_bef)<-c("CI_env1", "CI_env2")

ggplot() +
  geom_line(aes(x=1:100, y=GF_dist_bef$CI_env1) ,col="orange", lty=2,size=1.5)+
  geom_line(aes(x=1:100, y=GF_dist_bef$CI_env2) ,col="cyan", lty=2,size=1.5)+
  xlab("Location")+
  ylab("Cumulative Importance")


CI<-NULL
for(i in 1:nrow(GF_dist_bef)){
  for(j in 1:nrow(GF_dist_bef)){
    CI<-c(CI,dist(rbind(GF_dist_bef[i,],GF_dist_bef[j,])))
    #CI<-c(CI,CI_bf[j]-CI_bf[i])
  }
}

cg_df$D_CI<-CI


GFsel_dist_bef<-data.frame(cbind(gfTrans1_sel1,gfTrans1_sel2))
colnames(GFsel_dist_bef)<-c("CI_env1", "CI_env2")

CI_sel<-NULL
for(i in 1:nrow(GFsel_dist_bef)){
  for(j in 1:nrow(GFsel_dist_bef)){
    CI_sel<-c(CI_sel,dist(rbind(GFsel_dist_bef[i,],GFsel_dist_bef[j,])))
    #CI<-c(CI,CI_bf[j]-CI_bf[i])
  }
}

cg_df$D_CI_sel<-CI_sel


cg_df$FST<-as.vector(Pre_FST)
cg_df$FST_sel<-as.vector(Pre_FST_sel)


dat_filt<-cg_df[cg_df$Transplant%in%c("T11"),]
dat_sum<-cor(y=dat_filt$Fitness,x=dat_filt$D_CI, method = "pearson")

ggplot(data=dat_filt)+
  geom_smooth(aes(y=Fitness, x=D_CI), method="lm",se = F)+
  annotate("text",x=0.2,y=0.75,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
  geom_point(aes(x=D_CI, y=Fitness, col=Home), size=3)+
  labs(shape="Home", colour="Common\nGarden\nLocation")+
  scale_color_manual(values = colfunc(100))+
  ylab("Average relative fitness in common garden")+
  xlab("Absolute GF Offset (genome)")+
  ylim(0,1)+
  xlim(0,0.25)+
  ggtitle("Transplant to T11")+
  theme_classic()+
  theme(axis.text=element_text(size=12),axis.title=element_text(size=14))



dat_filt<-cg_df[cg_df$Transplant%in%c("T41"),]
dat_sum<-cor.test(y=dat_filt$Fitness,x=dat_filt$D_CI_sel, method = "pearson")

ggplot(data=dat_filt)+
  geom_smooth(aes(y=Fitness, x=D_CI_sel), method="lm",se = F)+
  annotate("text",x=0.3,y=0.75,label=paste("\u03C1 = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
  geom_point(aes(x=D_CI_sel, y=Fitness, col=Home), size=3)+
  labs(shape="Home", colour="Source Location")+
  scale_color_manual(values = colfunc(100))+
  ylab("Average relative fitness in common garden")+
  xlab("Absolute GF Offset (alleles)")+
  ylim(0,1)+
  # xlim(0,0.25)+
  ggtitle("Transplant to T41")+
  theme_classic()+
  theme(axis.text=element_text(size=12),axis.title=element_text(size=14))


#demes <-c("T1","T2","T11","T12","T21","T22","T9","T10","T19","T20","T29","T30")
demes <-c("T34","T35","T36","T37","T44","T45","T46","T47","T54","T55","T56","T57")
for (i in demes){
  dat_filt<-cg_df[cg_df$Transplant%in%c(i),]
  dat_sum<-cor.test(x=dat_filt$FST, y=dat_filt$Fitness, method = "pearson")

  print(ggplot(data=dat_filt)+
          geom_smooth(aes(x=FST, y=Fitness), method="lm",se = F)+
          annotate("text",x=0.006,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
          geom_point(aes(x=FST, y=Fitness, col=Home), size=3)+
          labs(shape="Home", colour="Common\nGarden\nLocation")+
          scale_color_manual(values = colfunc(100), guide=F)+
          ylab("Average relative fitness in common garden")+
          xlab("Pairwise FST (genome)")+
          ylim(0,1)+
          xlim(0,0.007)+
          ggtitle(paste("Transplant to",str_sub(i,2,3)))+
          theme_classic()+
          theme(axis.text=element_text(size=12),axis.title=element_text(size=14))
  )
}

#demes <-c("T1","T2","T11","T12","T21","T22","T9","T10","T19","T20","T29","T30")
demes <-c("T34","T35","T36","T37","T44","T45","T46","T47","T54","T55","T56","T57")

for (i in demes){
  dat_filt<-cg_df[cg_df$Transplant%in%c(i),]
  dat_sum<-cor.test(x=dat_filt$FST_sel, y=dat_filt$Fitness, method = "pearson")

  print(ggplot(data=dat_filt)+
          geom_smooth(aes(x=FST_sel, y=Fitness), method="lm",se = F)+
          annotate("text",x=0.03,y=0.95,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
          geom_point(aes(x=FST_sel, y=Fitness, col=Home), size=3)+
          labs(shape="Home", colour="Common\nGarden\nLocation")+
          scale_color_manual(values = colfunc(100), guide=F)+
          ylab("Average relative fitness in common garden")+
          xlab("Pairwise FST (allele)")+
          ylim(0,1)+
          xlim(0,0.05)+
          ggtitle(paste("Transplant to",str_sub(i,2,3)))+
          theme_classic()+
          theme(axis.text=element_text(size=12),axis.title=element_text(size=14))
  )
}


cg_ind<-read.table(paste("results/SLiM_output/CG_files/",seed,"_fitnessmat_ind.txt",sep=""),header=T)

colnames(cg_ind)<-rep(paste("ID",seq(1:10000),sep=""))
cg_ind$Transplant<-paste("T",seq(1:100),sep="")

cg_df_ind <- melt(cg_ind,  id.vars ="Transplant",  variable.name = "Individual",
              value.name = "Fitness")

cg_df_ind$Transplant<-factor(cg_df_ind$Transplant,levels=unique(cg_df_ind$Transplant))

Home<-NULL
for(i in 1:100){
  Home<-c(Home,rep(paste("H",i,sep=""),10000))
}
cg_df_ind$Home <- Home
cg_df_ind$Home<-factor(cg_df_ind$Home,levels=unique(cg_df_ind$Home))


CI<-NULL
for(i in 1:nrow(GF_dist_bef)){
  for(j in 1:nrow(GF_dist_bef)){
    CI<-c(CI,dist(rbind(GF_dist_bef[i,],GF_dist_bef[j,])))
    #CI<-c(CI,CI_bf[j]-CI_bf[i])
  }
}

CI_ind_rep<-NULL
for (i in seq(1,10000,100)){
  CI_ind_rep<-c(CI_ind_rep,rep(CI[i:(i+99)],100))
}

cg_df_ind$D_CI<-CI_ind_rep

GFsel_dist_bef<-data.frame(cbind(gfTrans1_sel1,gfTrans1_sel2))
colnames(GFsel_dist_bef)<-c("CI_env1", "CI_env2")

CI_sel<-NULL
for(i in 1:nrow(GFsel_dist_bef)){
  for(j in 1:nrow(GFsel_dist_bef)){
    CI_sel<-c(CI_sel,dist(rbind(GFsel_dist_bef[i,],GFsel_dist_bef[j,])))
    #CI<-c(CI,CI_bf[j]-CI_bf[i])
  }
}

CI_sel_ind_rep<-NULL
for (i in seq(1,10000,100)){
  CI_sel_ind_rep<-c(CI_sel_ind_rep,rep(CI_sel[i:(i+99)],100))
}

cg_df_ind$D_CI_sel<-CI_sel_ind_rep



dat_filt_ind<-cg_df_ind[cg_df_ind$Transplant%in%c("T1"),]

ggplot(data=dat_filt_ind)+
  geom_point(aes(x=D_CI, y=Fitness, col=Home), size=3)+
  geom_smooth(aes(y=Fitness, x=D_CI), method="lm",se = F)+
  labs(colour="Home Location")+
  scale_color_manual(values = colfunc(100))+
  #scale_shape_manual(values=c(15:115))+
  ylab("Average relative fitness in common garden")+
  xlab("Absolute GF Offset (genome)")+
  ylim(0,1)+
  xlim(0,0.25)+
  ggtitle("Transplant to T1")+
  theme_classic()+
  theme(axis.text=element_text(size=12),axis.title=element_text(size=14))




dat_filt<-cg_df_ind[cg_df_ind$Transplant%in%c("T11"),]
dat_sum<-cor.test(y=dat_filt$Fitness,x=dat_filt$D_CI_sel, method = "pearson")

ggplot(data=dat_filt)+
  annotate("text",x=0.35,y=0.75,label=paste("\u03C1 = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
  geom_point(aes(x=D_CI_sel, y=Fitness, col=Home), size=3)+
  geom_smooth(aes(y=Fitness, x=D_CI_sel), method="lm",se = F)+
  labs(shape="Home", colour="Source Location")+
  scale_color_manual(values = colfunc(100))+
  ylab("Individual relative fitness in common garden")+
  xlab("Absolute GF Offset (alleles)")+
  ylim(0,1)+
  xlim(0,0.45)+
  ggtitle("Transplant to T11")+
  theme_classic()+
  theme(axis.text=element_text(size=12),axis.title=element_text(size=14))

demes <-c("T34","T35","T36","T37","T44","T45","T46","T47","T54","T55","T56","T57")
for (i in demes){
  dat_filt<-cg_df_ind[cg_df_ind$Transplant%in%c(i),]
  dat_sum<-cor.test(y=dat_filt$Fitness,x=dat_filt$D_CI_sel, method = "pearson")

print(ggplot(data=dat_filt)+
  annotate("text",x=0.35,y=0.75,label=paste("r =",round(dat_sum$estimate[[1]],3),"\np-value =",signif(dat_sum$p.value,3)))+
  geom_point(aes(x=D_CI_sel, y=Fitness, col=Home), alpha = 0.2, size=3)+
  geom_smooth(aes(y=Fitness, x=D_CI_sel), method="lm",se = F)+
  labs(shape="Home", colour="Source Location")+
  scale_color_manual(values = colfunc(100), guide=F)+
  ylab("Individual relative fitness in common garden")+
  xlab("Absolute GF Offset (alleles)")+
  ylim(0,1)+
  xlim(0,0.45)+
  ggtitle(paste("Transplant to",i))+
  theme_classic()+
  theme(axis.text=element_text(size=12),axis.title=element_text(size=14))
  )
}


CHR_data <- data.frame (SNP = Pfst_pre_filt$LocusName , CHR=Linkage, BP= position1_filt_scaled, P=Pfst_pre_filt$FST)
# Plot !
manhattan(CHR_data , ylim=c(0,0.2),suggestiveline = F, genomewideline = F , logp=F, col=brewer.pal(5, "Set2") , main=plotTitle, ylab=expression(paste("Gen. offset (F"[ST],")")))

sel_pos<-position1_filt[position1_filt%in%unique(pos_T1)]

#FST outlier detection
Pfst_pre_filt_sel<-Pfst_pre_filt[which(Pfst_pre_filt$LocusName%in%select_pos),]
points(x=sel_pos, y=Pfst_pre_filt_sel$FST,col="red", pch=0.7)
