require(ggplot2)
require(cowplot)
require(qqman)
require(RColorBrewer)

colfunc<-colorRampPalette(c("orange","cyan"))
#plot(c(1:10),rep(1,10),col=colfunc(10),pch=19,cex=5) #Get color gradients for matrix image
#fit[1,210:309]<-fit[2,210:309]
plot(fit$Gen,fit$P19_env,type="l", main=plotTitle, xlab="Generations",ylab="Environmental variable",col="white",ylim=c(-5,10))
k=1
for(i in 209:308){
  lines(fit$Gen,fit[,i],lwd=2,col=colfunc(10)[k])
  k=k+1
}

#For neutral simulation
# plot(fit$Gen,fit$P19_env,type="l", main=plotTitle, xlab="Generations",ylab="Environmental variable",col="white",ylim=c(-5,10))
# k=1
# for(i in 218:317){
#   lines(fit$Gen,fit[,i],lwd=2,col=colfunc(10)[k])
#   k=k+1
# }
# lines(fit$Gen,fit$P19_hab, col="red")

#par(fig = c(0,1,0,1))
plot(fit$Gen,fit$P19_fit,type="l",main=plotTitle ,xlab="Generations",ylab="Relative Fitness",col="red",ylim=c(min(fit[,c(9:108)]),max(fit[,c(9:108)])))
x=1:10
y=1:10
for(i in x){
  for(j in y){
    k=((i - 1) + (j - 1) * 10 + 9)
    #print(k)
    lines(fit$Gen,fit[,k],col=colfunc(10)[i],lwd=2)
  }
}
lines(fit$Gen,fit$P19_fit, col="red")

#For neutral simulation 
# plot(fit$Gen,fit$P19_fit,type="l",main=plotTitle ,xlab="Generations",ylab="Relative Fitness",col="red")
# x=1:10
# y=1:10
# for(i in x){
#   for(j in y){
#     k=((i - 1) + (j - 1) * 10 + 18)
#     lines(fit$Gen,fit[,k],col=colfunc(10)[i])
#   }
# }
# lines(fit$Gen,fit$P19_fit, col="red")
# text(249900,0.78,"Location")
# par(fig = c(0.05,0.5, 0.07, 0.55), new = T)
# image(matrix(nrow=10,ncol=10,rep(1:10,10),byrow = F),col=colfunc(10),axes=F)+
#  abline(v=0.89,col="red")

#For mulilocus model
plot(fit$Gen,fit$P19_env1,type="l", main=plotTitle, xlab="Generations",ylab="Environmental optimum 1",col="white",ylim=c(-1,1))
k=1
for(i in 212:311){
  lines(fit$Gen,fit[,i],lwd=2,col=colfunc(10)[k])
  k=k+1
}

plot(fit$Gen,fit$P19_env2,type="l", main=plotTitle, xlab="Generations",ylab="Environmental optimum 2",col="white",ylim=c(-1,20))
deme=rep(1:10,each=10)
k=1
for(i in 312:411){
  lines(fit$Gen,fit[,i],lwd=2,col=colfunc(10)[deme[k]])
  k=k+1
}

plot(fit$Gen,fit$P19_freq,type="l", main=plotTitle, xlab="Generations",ylab="Mutant allele frequency",col="red",ylim=c(0,5))
x=1:10
y=1:10
for(i in x){
  for(j in y){
    k=((i - 1) + (j - 1) * 10 + 112)
    lines(fit$Gen,fit[,k],col=colfunc(10)[i])
  }
}
###################################################
#par(fig = c(0,1,0,1))
plot(fit$Gen,fit$P19_freq,type="l", main=plotTitle, xlab="Generations",ylab="Mutant allele frequency",col="red",ylim=c(0,1))
x=1:10
y=1:10
for(i in x){
  for(j in y){
    k=((i - 1) + (j - 1) * 10 + 109)
    lines(fit$Gen,fit[,k],col=colfunc(10)[i])
  }
}
lines(fit$Gen,fit$P19_freq, col="red")
#abline(h=1)

ggplot()+
  geom_boxplot(aes(x=fitt[fitt$Type=="Env",]$Location,y=fitt[fitt$Type=="Fit",]$Gen301,col=fitt[fitt$Type=="Fit",]$Location))+
  scale_colour_manual(values = colfunc(10))+
  xlab("Environment type")+
  ylab("Relative fitness")+
  theme_classic()+
  theme(legend.position = "none")

ggplot()+
  geom_boxplot(aes(x=fitt[fitt$Type=="Env",]$Location,y=fitt[fitt$Type=="Fit",]$Gen601,col=fitt[fitt$Type=="Fit",]$Location))+
  scale_colour_manual(values = colfunc(10))+
  xlab("Environment type")+
  ylab("Relative fitness")+
  theme_classic()+
  theme(legend.position = "none")

ggplot()+
  geom_boxplot(aes(x=fitt[fitt$Type=="Env",]$Location,y=fitt[fitt$Type=="Freq",]$Gen301,col=fitt[fitt$Type=="Freq",]$Location))+
  scale_colour_manual(values = colfunc(10))+
  xlab("Environment type")+
  ylab("Allele frequency of derived allele under selection")+
  theme_classic()+
  theme(legend.position = "none")

ggplot()+
  geom_boxplot(aes(x=fitt[fitt$Type=="Env",]$Location,y=fitt[fitt$Type=="Freq",]$Gen601,col=fitt[fitt$Type=="Freq",]$Location))+
  scale_colour_manual(values = colfunc(10))+
  xlab("Environment type")+
  ylab("Allele frequency of allele under selection")+
  theme_classic()+
  theme(legend.position = "none")

ggplot()+
  geom_boxplot(aes(x=fitt[fitt$Type=="Env",]$Location,y=Pop_afreq1$M224999,col=fitt[fitt$Type=="Freq",]$Location))+
  scale_colour_manual(values = colfunc(10))+
  xlab("Environment type")+
  ylab("Allele frequency of allele under selection")+
  theme_classic()+
  theme(legend.position = "none")

ggplot()+
  geom_boxplot(aes(x=fitt[fitt$Type=="Env",]$Location,y=Pop_afreq2$M224999,col=fitt[fitt$Type=="Freq",]$Location))+
  scale_colour_manual(values = colfunc(10))+
  xlab("Environment type")+
  ylab("Allele frequency of allele under selection")+
  theme_classic()+
  theme(legend.position = "none")

fitt[fitt$Type=="Freq",]$Gen301 - Pop_afreq1$M224999
fitt[fitt$Type=="Freq",]$Gen601 - Pop_afreq2$M224999

#For neutral simulation
# plot(fit$Gen,fit$P19_freq,type="l", main=plotTitle, xlab="Generations",ylab="Mutant allele frequency",col="red",ylim=c(0,1))
# x=1:10
# y=1:10
# for(i in x){
#   for(j in y){
#     k=((i - 1) + (j - 1) * 10 + 118)
#     lines(fit$Gen,fit[,k],col=colfunc(10)[i])
#   }
# }
# lines(fit$Gen,fit$P19_freq, col="red")

ggplot() + 
  geom_line(aes(x=envPop$envSelect, y=gfM2Trans1$C.Imp_M2_before), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="Cumulative Importance", x="Environment") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  ggtitle(paste(plotTitle,"Gen. 300")) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))

ggplot() + 
  geom_line(aes(x=envPop$envSelect, y=gfM1Trans1$C.Imp_M1_before), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="Cumulative Importance", x="Environment") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  ggtitle(paste(plotTitle,", Gen. 300")) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))

 ggplot() + 
  geom_line(aes(x=envPop$envSelect, y=gfTrans1$C.Imp_genome_before), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="GF Cumulative Importance", x="Environment") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  ggtitle(plotTitle) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))

# plot(Summary_Pop$Pop,Summary_Pop$D_C.Imp_genome,xlab="Population",ylab="C.Imp.Genome")
# plot(Summary_Pop$Pop,Summary_Pop$D_C.Imp_M2,xlab="Population",ylab="C.Imp.M2")
# plot(Summary_Pop$Pop,Summary_Pop$F_ST_M2_bef.aft.,xlab="Population",ylab="FST bef.aft. M2")
# plot(Summary_Pop$Pop,Summary_Pop$F_ST_genome_bef.aft.,xlab="Population",ylab="FST bef.aft. Genome")

#par(mfrow=c(2,2),bty="l",mar=c(4,4,1,1))
plot(Summary_Pop$F_ST_genome_bef.aft.,Summary_Pop$D_C.Imp_genome,col=colfunc(10),pch=19, cex=2,ylab="GF Gen offset Genome",xlab=expression(paste("True Gen offset (F"[ST]," Genome)")))
#abline(lm(Summary_Pop$D_C.Imp_genome~Summary_Pop$F_ST_genome_bef.aft.),col="red")
rho_test<-cor.test(Summary_Pop$D_C.Imp_genome,Summary_Pop$F_ST_genome_bef.aft., method = "spearman" )
text(adj=0, max(Summary_Pop$F_ST_genome_bef.aft.)-max(Summary_Pop$F_ST_genome_bef.aft.)*0.15,max(Summary_Pop$D_C.Imp_genome)-max(Summary_Pop$D_C.Imp_genome)*0.1,paste("Rho = ",round(rho_test$estimate,2), "\np-value = ",round(rho_test$p.value,2), sep=""))

# plot(Summary_Pop$Rel_Fit_diff,Summary_Pop$D_C.Imp_genome,ylab="GF Gen offset Genome",xlab="True Fitness Offset")
# abline(lm(Summary_Pop$D_C.Imp_genome~Summary_Pop$Rel_Fit_diff),col="red")
# rho_test<-cor.test(Summary_Pop$D_C.Imp_genome,Summary_Pop$Rel_Fit_diff, method = "spearman" )
# text(adj=0, -0.4,0.09,paste("Rho = ",round(rho_test$estimate,4), "\np-value = ",round(rho_test$p.value,4), sep=""))
# 
# plot(Summary_Pop$F_ST_M2_bef.aft.,Summary_Pop$D_C.Imp_M2,ylab="GF Gen offset M2",xlab=expression(paste("True Gen offset (F"[ST]," M2)")))
# abline(lm(Summary_Pop$D_C.Imp_M2~Summary_Pop$F_ST_M2_bef.aft.),col="red")
# rho_test<-cor.test(Summary_Pop$D_C.Imp_M2,Summary_Pop$F_ST_M2_bef.aft., method = "spearman" )
# text(adj=0, 0,0.8,paste("Rho = ",round(rho_test$estimate,4), "\np-value = ",round(rho_test$p.value,16), sep=""))
# 
# plot(Summary_Pop$Rel_Fit_diff,Summary_Pop$D_C.Imp_M2,ylab="GF Gen offset M2",xlab="True Fitness Offset")
# #abline(lm(Summary_Pop$D_C.Imp_M2~Summary_Pop$Rel_Fit_diff),col="red")
# rho_test<-cor.test(Summary_Pop$D_C.Imp_M2,Summary_Pop$Rel_Fit_diff, method = "spearman" )
# text(adj=0,-0.4,0.8,paste("Rho = ",round(rho_test$estimate,4), "\np-value = ",round(rho_test$p.value,4), sep=""))
# 
# plot(Summary_Pop$Rel_Fit_diff,Summary_Pop$F_ST_genome_bef.aft.,ylab=expression(paste("True Gen offset (F"[ST]," Genome)")),xlab="True Fitness Offset")
# abline(lm(Summary_Pop$F_ST_genome_bef.aft.~Summary_Pop$Rel_Fit_diff),col="red")
# rho_test<-cor.test(Summary_Pop$F_ST_genome_bef.aft.,Summary_Pop$Rel_Fit_diff, method = "spearman" )
# text(adj=0,-0.3,0.038,paste("Rho = ",round(rho_test$estimate,4), "\np-value = < 2.2e-16", sep=""))
# 
# plot(Summary_Pop$Rel_Fit_diff,Summary_Pop$F_ST_M2_bef.aft.,ylab=expression(paste("True Gen offset (F"[ST]," M2)")),xlab="True Fitness Offset")
# abline(lm(Summary_Pop$F_ST_M2_bef.aft.~Summary_Pop$Rel_Fit_diff),col="red")
# rho_test<-cor.test(Summary_Pop$F_ST_M2_bef.aft.,Summary_Pop$Rel_Fit_diff, method = "spearman" )
# text(adj=0,-0.2,0.9,paste("Rho = ",round(rho_test$estimate,4), "\np-value = ",round(rho_test$p.value,16), sep=""))

rho_test<-cor.test(Summary_Pop$Rel_Fit_diff,Summary_Pop$F_ST_genome_bef.aft., method = "spearman" )
ggplot(data=Summary_Pop,aes(x=Summary_Pop$F_ST_genome_bef.aft.,y=Summary_Pop$Rel_Fit_diff,color=Summary_Pop$X,size=2))+
  geom_point()+
  ylab("True Fitness Offset")+
  xlab(expression(paste("True Gen offset (F"[ST]," Genome)")))+
  scale_colour_gradientn(colours = colfunc(10))+
  theme_classic(base_size = 16)+
  theme(legend.position = "none")+
  geom_smooth(method='lm',col="red",size=1,se=F)+
  annotate("text", x = abs(max(Summary_Pop$F_ST_genome_bef.aft.)-max(Summary_Pop$F_ST_genome_bef.aft.)*0.1), y = max(Summary_Pop$Rel_Fit_diff)-max(Summary_Pop$Rel_Fit_diff)*0.1, label = paste("Rho = ",round(rho_test$estimate,2), "\np-value = ",round(rho_test$p.value,4), sep=""))

rho_test<-cor.test(Summary_Pop$Rel_Fit_diff,Summary_Pop$F_ST_M2_bef.aft., method = "spearman" )
ggplot(data=Summary_Pop,aes(x=Summary_Pop$F_ST_M2_bef.aft.,y=Summary_Pop$Rel_Fit_diff,color=Summary_Pop$X,size=2))+
  geom_point()+
  ylab("True Fitness Offset")+
  xlab(expression(paste("True Gen offset (F"[ST]," M2)")))+
  scale_colour_gradientn(colours = colfunc(10))+
  theme_classic(base_size = 16)+
  theme(legend.position = "none")+
  geom_smooth(method='lm',col="red",size=1,se=F)+
  annotate("text", x = min(Summary_Pop$F_ST_M2_bef.aft.)+0.1, y = min(Summary_Pop$Rel_Fit_diff)+0.1, label = paste("Rho = ",round(rho_test$estimate,2), "\np-value = ",round(rho_test$p.value,12), sep=""))


rho_test<-cor.test(Summary_Pop$D_C.Imp_genome,Summary_Pop$F_ST_genome_bef.aft., method = "spearman" )
ggplot(data=Summary_Pop,aes(x=Summary_Pop$F_ST_genome_bef.aft.,y=Summary_Pop$D_C.Imp_genome,color=Summary_Pop$X,size=2))+
  geom_point()+
  ylab("GF offset Genome")+
  xlab(expression(paste("Gen. offset (F"[ST]," Genome)")))+
  scale_colour_gradientn(colours = colfunc(10))+
  theme_classic(base_size = 16)+
  theme(legend.position = "none")+
  geom_smooth(method='lm',col="red",size=1,se=F)+
  annotate("text", x = max(Summary_Pop$F_ST_genome_bef.aft.)-max(Summary_Pop$F_ST_genome_bef.aft.)*0.05, y = max(Summary_Pop$D_C.Imp_genome)-max(Summary_Pop$D_C.Imp_genome)*0.1, label = paste("Rho = ",round(rho_test$estimate,3), "\np-value = ",round(rho_test$p.value,3), sep=""))

rho_test<-cor.test(Summary_Pop$D_C.Imp_M2,Summary_Pop$F_ST_M2_bef.aft., method = "spearman" )
ggplot(data=Summary_Pop,aes(x=Summary_Pop$F_ST_M2_bef.aft.,y=Summary_Pop$D_C.Imp_M2,color=Summary_Pop$X,size=2))+
  geom_point()+
  ylab("GF Gen offset of allele under selection")+
  xlab(expression(paste("True Gen offset (F"[ST]," of allele under selection)")))+
  scale_colour_gradientn(colours = colfunc(10))+
  theme_classic(base_size = 16)+
  theme(legend.position = "none")+
  geom_smooth(method='lm',col="red",size=1,se=F)+
  annotate("text", x = min(Summary_Pop$F_ST_M2_bef.aft.)+min(Summary_Pop$F_ST_M2_bef.aft.)+0.1, y = max(Summary_Pop$D_C.Imp_M2)-max(Summary_Pop$D_C.Imp_M2)*0.1, label = paste("Rho = ",round(rho_test$estimate,2), "\np-value = ",round(rho_test$p.value,11), sep=""))

rho_test<-cor.test(Summary_Pop$D_C.Imp_genome,Summary_Pop$Rel_Fit_diff, method = "spearman" )
ggplot(data=Summary_Pop,aes(x=Summary_Pop$Rel_Fit_diff,y=Summary_Pop$D_C.Imp_genome,color=Summary_Pop$X,size=2))+
  geom_point()+
  ylab("GF Gen offset Genome")+
  xlab("True Fitness Offset")+
  scale_colour_gradientn(colours = colfunc(10))+
  theme_classic(base_size = 16)+
  theme(legend.position = "none")+
  annotate("text", x = min(Summary_Pop$Rel_Fit_diff)-min(Summary_Pop$Rel_Fit_diff)*0.15, y = max(Summary_Pop$D_C.Imp_genome)-max(Summary_Pop$D_C.Imp_genome)*0.1, label = paste("Rho = ",round(rho_test$estimate,2), "\np-value = ",round(rho_test$p.value,7), sep=""))

rho_test<-cor.test(Summary_Pop$D_C.Imp_genome,abs(Summary_Pop$Rel_Fit_diff), method = "spearman" )
ggplot(data=Summary_Pop,aes(x=abs(Summary_Pop$Rel_Fit_diff),y=Summary_Pop$D_C.Imp_genome,color=Summary_Pop$X,size=2))+
  geom_point()+
  ylab("GF Gen offset Genome")+
  xlab("True Fitness Offset (absolute)")+
  scale_colour_gradientn(colours = colfunc(10))+
  theme_classic(base_size = 16)+
  theme(legend.position = "none")+
  annotate("text", x = abs(min(Summary_Pop$Rel_Fit_diff)-min(Summary_Pop$Rel_Fit_diff))+0.4, y = max(Summary_Pop$D_C.Imp_genome)-max(Summary_Pop$D_C.Imp_genome)*0.1, label = paste("Rho = ",round(rho_test$estimate,2), "\np-value = ",round(rho_test$p.value,2), sep=""))

rho_test<-cor.test(Summary_Pop$D_C.Imp_M2,Summary_Pop$Rel_Fit_diff, method = "spearman" )
ggplot(data=Summary_Pop)+
  geom_point(aes(x=Summary_Pop$Rel_Fit_diff,y=Summary_Pop$D_C.Imp_M2,color=Summary_Pop$X,size=2))+
  ylab("GF Gen offset of allele under selection")+
  xlab("True Fitness Offset")+
  scale_colour_gradientn(colours = colfunc(10))+
  theme_classic(base_size = 16)+
  theme(legend.position = "none")+
  annotate("text", x = min(Summary_Pop$Rel_Fit_diff)-min(Summary_Pop$Rel_Fit_diff)*0.15, y = max(Summary_Pop$D_C.Imp_M2)-max(Summary_Pop$D_C.Imp_M2)*0.1, label = paste("Rho = ",round(rho_test$estimate,2), "\np-value = ",round(rho_test$p.value,2), sep=""))

rho_test<-cor.test(Summary_Pop$D_C.Imp_M2,abs(Summary_Pop$Rel_Fit_diff), method = "spearman" )
ggplot(data=Summary_Pop,aes(x=abs(Summary_Pop$Rel_Fit_diff),y=Summary_Pop$D_C.Imp_M2,color=Summary_Pop$X))+
  geom_point(aes(size=2))+
  ylab("GF Gen offset of allele under selection")+
  xlab("True Fitness Offset (absolute)")+
  scale_colour_gradientn(colours = colfunc(10))+
  theme_classic(base_size = 16)+
  theme(legend.position = "none")+
  geom_smooth(method='lm',col="red",size=1,se=F)+
  annotate("text", x = min(Summary_Pop$Rel_Fit_diff)-min(Summary_Pop$Rel_Fit_diff)+0.4, y = max(Summary_Pop$D_C.Imp_M2)-max(Summary_Pop$D_C.Imp_M2)*0.1, label = paste("Rho = ",round(rho_test$estimate,2), "\np-value = ",round(rho_test$p.value,9), sep=""))

CHR_data <- data.frame (SNP = Pfst_pre_filt$LocusName , CHR=Linkage, BP= position1_filt_scaled, P=Pfst_pre_filt$FST)
# Plot !
manhattan(CHR_data , suggestiveline = F, genomewideline = F , logp=F, col=brewer.pal(5, "Set2") , cex=0.7, main=plotTitle, ylab=expression(paste("Gen. offset (F"[ST],")")))
abline(h=0.15,col="red")
#FST outlier detection

#For neutral simulation, plot deme size vs GF Offset
rho_test<-cor.test(Summary_Pop$D_C.Imp_genome,Summary_Pop$Pop_size, method = "spearman" )
ggplot(data=Summary_Pop,aes(x=Summary_Pop$D_C.Imp_genome,y=Summary_Pop$Pop_size,color=Summary_Pop$X,size=2))+
  geom_point()+
  ylab("Deme size")+
  xlab(expression(paste("GF Gen offset Genome")))+
  scale_colour_gradientn(colours = colfunc(10))+
  theme_classic(base_size = 16)+
  ylim(10,220)+
  theme(legend.position = "none")+
  geom_smooth(method='lm',col="red",size=1,se=F)+
  annotate("text", x = max(Summary_Pop$D_C.Imp_genome)-max(Summary_Pop$D_C.Imp_genome)*0.15, y = min(Summary_Pop$Pop_size)+min(Summary_Pop$Pop_size)*1.5, label = paste("Rho = ",round(rho_test$estimate,3), "\np-value = ",round(rho_test$p.value,22), sep=""))

