#Visualizing multilocus data
require(ggplot2)
require(cowplot)
require(qqman)
require(RColorBrewer)
require(colorspace)
require(dplyr)
require(reshape2)

setwd("E:/Research_AJL/TTT_Offset_Vulnerability_GF_Sims/")

colfunc<-colorRampPalette(c("orange","cyan"))

colour1 = rep(colfunc(10),10)
colour1_transparent <- adjustcolor(colour1, alpha.f = 0.3) 
par(mfrow=c(2,1))
par(mar = c(1,4,4,0))
plot(fit$Gen,fit$P19_phen1,type="l", main=plotTitle, xlab="",xaxt="n", ylab="Phenotype 1",col="white",ylim=c(-1,1))
k=1
for(i in which(fitt$Type=="Phen1")){
  lines(fit$Gen,fit[,i+13],lwd=2,col=colour1[k])
  #lines(fit$Gen,fit[,i+14],lwd=2,col=colour1[k])
  k=k+1
}
par(mar = c(5,4,0,0))
colour2 = rep(colfunc(10), each=10)
colour2_transparent <- adjustcolor(colour2, alpha.f = 0.3) 
plot(fit$Gen,fit$P19_phen1,type="l", main=NULL, xlab="Generations",ylab="Phenotype 2",col="white",ylim=c(-1,1))
k=1
for(i in which(fitt$Type=="Phen2")){
  lines(fit$Gen,fit[,i+13],lwd=2,col=colour2[k])
  #lines(fit$Gen,fit[,i+14],lwd=2,col=colour2[k])
  k=k+1
}

#dev.off()

par(mfrow=c(2,1))
par(mar = c(1,4,4,0))
plot(fit$Gen,fit$P19_env1,type="l", main=plotTitle,  xlab="",xaxt="n",ylab="Environment 1",col="white",ylim=c(-1,1))
k=1
for(i in which(fitt$Type=="Env1")){
  lines(fit$Generation,fit[,i+13],lwd=2,col=colour1[k])
  #lines(fit$Generation,fit[,i+14],lwd=2,col=colour1[k])
  k=k+1
}
par(mar = c(5,4,0,0))
plot(fit$Generation,fit$P1_env2,type="l", main=NULL, xlab="Generations",ylab="Environment 2",col="white",ylim=c(-1,1))
color=rep(colfunc(10), each=10)
k=1
for(i in which(fitt$Type=="Env2")){
  lines(fit$Generation,fit[,(i+13)],lwd=2,col=colour2[k])
  k=k+1
}

dev.off()

plot(fit$Gen,fit$P19_fit,type="l",main=plotTitle ,xlab="Generations",ylab="Absolute Fitness",col="red", ylim=c(0.6,1))
x=1:10
y=1:10
for(i in x){
  for(j in y){
    k=((i - 1) + (j - 1) * 10 + 13)
    #print(paste("k = ", k))
    lines(fit$Gen,fit[,k],col=colour1[i])
  }
}

#GF visualization per environment
A <- ggplot() +
  geom_line(aes(x=envTab[,1], y=gfTrans1e1$C.Imp_genome_envPop1), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  ylab("Cumulative Importance\n(Genome)")+
  xlab("") +
  #theme(axis.title.y = element_text(margin = margin(t = 50, r = 0, b = 0, l = 50)))+
  #labs(y="Cumulative Importance\n(Genome)", x="Environment 1") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  ylim(-0.003,0.16)+
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle(plotTitle) +
    #ggtitle(paste(plotTitle,", all alleles")) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

B <- ggplot() +
  geom_line(aes(x=envTab[,2], y=gfTrans1e2$C.Imp_genome_envPop2), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="Cumulative Importance\n(Genome)\n", x="") +
  #labs(y="Cumulative Importance\n(Genome)", x="Environment 2") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  #ggtitle(plotTitle) +
  scale_y_continuous(position = "right", sec.axis = sec_axis(~., labels = NULL),limits = c(-0.002,0.16)) +
  theme(plot.title = element_text(size=14, face="bold.italic"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

#GF visualization per environment with the just alleles under selection
C <- ggplot() +
  geom_line(aes(x=envTab[,1], y=gfTrans1_sel1$C.Imp_causal_envPop1), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="\n(Causal)", x="Environment 1") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  ylim(-0.003,0.16)+
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  #ggtitle(paste(plotTitle,", alleles under selection")) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))

D <- ggplot() +
  geom_line(aes(x=envTab[,2], y=gfTrans1_sel2$C.Imp_causal_envPop2), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  ylab("\n(Causal)\n") +
  xlab("Environment 2")+
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  #theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  scale_y_continuous(position = "right", sec.axis = sec_axis(~., labels = NULL),limits = c(-0.002,0.16)) +
  #ggtitle(plotTitle) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))

title <- ggdraw() + draw_label(plotTitle, fontface='bold')
plot_grid(title,NULL,A,B,C,D,nrow=3,ncol=2,rel_heights = c(1,3,3))

#Neutral environments
A <- ggplot() +
  geom_line(aes(x=envTab[,3], y=gfTrans1e3$C.Imp_genome_fakeEnv1), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  ylab("Cumulative Importance\n(Genome)")+
  xlab("") +
  #theme(axis.title.y = element_text(margin = margin(t = 50, r = 0, b = 0, l = 50)))+
  #labs(y="Cumulative Importance\n(Genome)", x="Environment 1") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  ylim(-0.003,0.16)+
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle(plotTitle) +
  #ggtitle(paste(plotTitle,", all alleles")) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

B <- ggplot() +
  geom_line(aes(x=envTab[,4], y=gfTrans1e4$C.Imp_genome_fakeEnv2), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="Cumulative Importance\n(Genome)\n", x="") +
  #labs(y="Cumulative Importance\n(Genome)", x="Environment 2") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  #ggtitle(plotTitle) +
  scale_y_continuous(position = "right", sec.axis = sec_axis(~., labels = NULL),limits = c(-0.002,0.16)) +
  theme(plot.title = element_text(size=14, face="bold.italic"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

#GF visualization per environment with the just alleles under selection
C <- ggplot() +
  geom_line(aes(x=envTab[,3], y=gfTrans1_sel3$C.Imp_causal_fakeEnv1), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  labs(y="\n(Causal)", x="Neutral Env. 1") +
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  ylim(-0.002,0.16)+
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  #ggtitle(paste(plotTitle,", alleles under selection")) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))

D <- ggplot() +
  geom_line(aes(x=envTab[,4], y=gfTrans1_sel4$C.Imp_causal_fakeEnv2), colour=rgb(0,0,0,0.4), lwd=1.5) +
  #facet_grid(. ~ strSel) +
  ylab("\n(Causal)\n") +
  xlab("Neutral Env. 2")+
  #geom_line(data=cImpMAF.neut, aes(x=x, y=y),
  #          colour=rgb(0,0,1, 0.75), lwd=1) +
  #geom_line(data=cImpMAF.sel, aes(x=x, y=y),
  #          colour=rgb(1,0,0, 0.75), lwd=1) +
  #geom_line(data=cImpMAF, aes(x=x, y=y),
  #          colour=rgb(0,0,0, 0.75), lwd=1) +
  #theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #ggtitle("Lost > 5") +
  scale_y_continuous(position = "right", sec.axis = sec_axis(~., labels = NULL),limits = c(-0.002,0.16)) +
  #ggtitle(plotTitle) +
  #scale_x_continuous(limits=c(-4,7))+
  theme(plot.title = element_text(size=14, face="bold.italic"))

title <- ggdraw() + draw_label(plotTitle, fontface='bold')
plot_grid(title,NULL,A,B,C,D,nrow=3,ncol=2,rel_heights = c(1,3,3))


## Visualize CI for individual alleles
cImp1sel <- cImp1[cImp1$allele%in%select_pos,]
A <- ggplot() +
  geom_line(data=cImp1, aes(x=x , y=y, group=allele), color="grey", lwd=1.5) +
  geom_line(data=cImp1sel, aes(x=x , y=y, group=allele),color="red",lwd=1.5) +
  ylab("Cumulative Importance")+
  xlab("Environment 1") +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  ylim(0.0,0.6)

cImp2sel <- cImp2[cImp2$allele%in%select_pos,]
B <- ggplot() +
  geom_line(data=cImp2, aes(x=x , y=y, group=allele), color="grey", lwd=1.5) +
  geom_line(data=cImp2sel, aes(x=x , y=y, group=allele),color="red",lwd=1.5) +
  ylab("Cumulative Importance")+
  xlab("Environment 2") +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  ylim(0.0,0.6)+
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

cImp3sel <- cImp3[cImp3$allele%in%select_pos,]
C <- ggplot() +
  geom_line(data=cImp3, aes(x=x , y=y, group=allele), color="grey", lwd=1.5) +
  geom_line(data=cImp3sel, aes(x=x , y=y, group=allele),color="red",lwd=1.5) +
  ylab("Cumulative Importance")+
  xlab("Neutral Env. 1") +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  ylim(0.0,0.6)

cImp4sel <- cImp4[cImp4$allele%in%select_pos,]
D <- ggplot() +
  geom_line(data=cImp4, aes(x=x , y=y, group=allele), color="grey", lwd=1.5) +
  geom_line(data=cImp4sel, aes(x=x , y=y, group=allele),color="red",lwd=1.5) +
  ylab("Cumulative Importance")+
  xlab("Neutral Env. 2") +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  ylim(0.0,0.6)+
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

title <- ggdraw() + draw_label(plotTitle, fontface='bold')
#plot_grid(title,NULL,A,B,nrow=2,ncol=2,rel_heights = c(1,3))
plot_grid(title,NULL,A,B,C,D,nrow=3,ncol=2,rel_heights = c(1,3,3))


#plot the allele frequencies of the ten alleles with the highest CI values
top_al1<-unique(cImp1[order(cImp1$y,decreasing = T),]$allele)[1:10]
length(unique(cImp1$allele))

alf1<-Pop_afreq1[,colnames(Pop_afreq1)%in%top_al1]

alf1$Env1 <- envPop1$envPop1

alf1b <- melt(alf1, id = "Env1", variable.name = "Allele")

alf1c <- alf1b %>% 
  group_by(Env1,Allele) %>% 
  summarise(average = mean(value))

A <- 
  ggplot() +
  geom_line(data=alf1c, aes(x=Env1, y=average, linetype=Allele, color=Allele),lwd=1) +
  ylab("Allele frequency (average)")+
  xlab("Environment 1") +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  #scale_color_discrete_qualitative(palette = "Pastel")+
  #geom_text(aes(x=rep(-0.75,10),y= c(seq(0.5,0.3,length.out =10)), label =top_al1))+
  #annotate("text",x = -1, y =0.5, label=paste(top_al1))+
  ylim(0.0,1)

top_al2<-unique(cImp2[order(cImp2$y,decreasing = T),]$allele)[1:10]
length(unique(cImp2$allele))

alf2<-Pop_afreq1[,colnames(Pop_afreq1)%in%top_al2]

alf2$Env2 <- envPop2$envPop2

alf2b <- melt(alf2, id = "Env2", variable.name = "Allele")

alf2c <- alf2b %>% 
  group_by(Env2,Allele) %>% 
  summarise(average = mean(value))

B <- ggplot() +
  geom_line(data=alf2c, aes(x=Env2, y=average, linetype=Allele, color=Allele),lwd=1) +
  ylab("")+
  xlab("Environment 2") +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  # theme(axis.text.y = element_text(size = 16, colour = "grey60"),
  #       axis.title.y = element_text(size=24, vjust=1)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16))+
  ylim(0.0,1)+
  theme(plot.title = element_text(size=14, face="bold.italic"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())


title <- ggdraw() + draw_label(plotTitle, fontface='bold')
plot_grid(title,NULL,A,B,nrow=2,ncol=2,rel_heights = c(1,3), rel_widths = c(1.1,1))



#plot the allele frequencies of all the considered adapted alleles

alf1s<-alFreq_sel

alf1s$Env1 <- envPop1$envPop1

alf1sb <- melt(alf1s, id = "Env1", variable.name = "Allele")

alf1sc <- alf1sb %>% 
  group_by(Env1,Allele) %>% 
  summarise(average = mean(value))

colfunc<-colorRampPalette(c("orange","cyan"))

colour1 = colfunc(length(unique(alf1sc$Allele)))
colour1_transparent <- adjustcolor(colour1, alpha.f = 0.5) 

A <- 
  ggplot() +
  geom_line(data=alf1sc, aes(x=Env1, y=average, color=Allele),lwd=2) +
  ylab("Adapted allele frequency (average)")+
  xlab("Environment 1") +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16)) +
  scale_colour_manual(values=colour1_transparent)+
  #scale_color_discrete_qualitative(palette = "Pastel")+
  #geom_text(aes(x=rep(-0.75,10),y= c(seq(0.5,0.3,length.out =10)), label =top_al1))+
  #annotate("text",x = -1, y =0.5, label=paste(top_al1))+
  guides(color = FALSE) +
  ylim(0.0,1)

alf2s<-alFreq_sel

alf2s$Env2 <- envPop2$envPop2

alf2sb <- melt(alf2s, id = "Env2", variable.name = "Allele")

alf2sc <- alf2sb %>% 
  group_by(Env2,Allele) %>% 
  summarise(average = mean(value))

colour2 = colfunc(length(unique(alf2sc$Allele)))
colour2_transparent <- adjustcolor(colour2, alpha.f = 0.5) 

B <- ggplot() +
  geom_line(data=alf2sc, aes(x=Env2, y=average, color=Allele),lwd=2) +
  ylab("")+
  xlab("Environment 2") +
  theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 18, colour = "grey60"),
        axis.title.x = element_text(size=24)) +
  theme(axis.text.y = element_text(size = 16, colour = "grey60"),
        axis.title.y = element_text(size=24, vjust=1)) +
  theme(strip.text = element_text(size=16))+
  scale_colour_manual(values=colour2_transparent)+
  guides(color = FALSE) +
  theme(plot.title = element_text(size=14, face="bold.italic"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())+
  ylim(0.0,1)


title <- ggdraw() + draw_label(plotTitle, fontface='bold')
plot_grid(title,NULL,A,B,nrow=2,ncol=2,rel_heights = c(1,3), rel_widths = c(1.1,1))


#write.csv(case, "results/R_results/Case1_corr.csv", row.names = F)
# Env_dist<-data.frame(cbind(envPop1,envPop2))
# colnames(Env_dist)<-c("Env1", "Env2")
# 
# DI<-NULL
# for(i in 1:nrow(Env_dist)){
#   for(j in 1:nrow(Env_dist)){
#     DI<-c(DI,dist(rbind(Env_dist[i,],Env_dist[j,])))
#     #CI<-c(CI,CI_bf[j]-CI_bf[i])
#   }
# }
# 
# cg_df$DI<-DI

demes <-unique(cg_df$Transplant)
FST_plots <- vector('list', length(demes))
max(cg_df[!is.na(cg_df$FST),]$FST)
min(cg_df[!is.na(cg_df$FST),]$FST)
for (i in seq_along(demes)) {
  dat_filt<-cg_df[cg_df$Transplant%in%demes[i],]
  dat_sum<-cor.test(x=dat_filt$FST, y=dat_filt$Fitness, method = "pearson")
  FST_plots[[i]] <- local({
    i <- i
    p1 <- ggplot(data=dat_filt)+
      geom_smooth(aes(x=FST, y=Fitness), method="lm",se = F)+
      annotate("text",x=0.002,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
      geom_point(aes(x=FST, y=Fitness, col=Home), size=3)+
      labs(shape="Home", colour="Common\nGarden\nLocation")+
      scale_color_manual(values = colfunc(100), guide=F)+
      # ylab("Average relative fitness in common garden")+
      # xlab("Pairwise FST (genome)")+
      ylab("")+
      xlab("")+
      ylim(0,1)+
      xlim(-0.003,0.008)+
      ggtitle(paste("Transplant to",str_sub(demes[i],2,4)))+
      theme_classic()+
      theme(axis.text=element_text(size=12),axis.title=element_text(size=14)
      )
    print(p1)
  })
}

plot_grid(FST_plots[[1]],FST_plots[[2]],FST_plots[[3]],FST_plots[[4]],FST_plots[[5]],FST_plots[[6]],FST_plots[[7]],FST_plots[[8]],FST_plots[[9]],FST_plots[[10]],
          FST_plots[[11]],FST_plots[[12]],FST_plots[[13]],FST_plots[[14]],FST_plots[[15]],FST_plots[[16]],FST_plots[[17]],FST_plots[[18]],FST_plots[[19]],FST_plots[[20]],
          FST_plots[[21]],FST_plots[[22]],FST_plots[[23]],FST_plots[[24]],FST_plots[[25]],FST_plots[[26]],FST_plots[[27]],FST_plots[[28]],FST_plots[[29]],FST_plots[[30]],
          FST_plots[[31]],FST_plots[[32]],FST_plots[[33]],FST_plots[[34]],FST_plots[[35]],FST_plots[[36]],FST_plots[[37]],FST_plots[[38]],FST_plots[[39]],FST_plots[[40]],
          FST_plots[[41]],FST_plots[[42]],FST_plots[[43]],FST_plots[[44]],FST_plots[[45]],FST_plots[[46]],FST_plots[[47]],FST_plots[[48]],FST_plots[[49]],FST_plots[[50]],
          nrow=5)

plot_grid(FST_plots[[51]],FST_plots[[52]],FST_plots[[53]],FST_plots[[54]],FST_plots[[55]],FST_plots[[56]],FST_plots[[57]],FST_plots[[58]],FST_plots[[59]],FST_plots[[60]],
          FST_plots[[66]],FST_plots[[62]],FST_plots[[63]],FST_plots[[64]],FST_plots[[65]],FST_plots[[66]],FST_plots[[67]],FST_plots[[68]],FST_plots[[69]],FST_plots[[70]],
          FST_plots[[71]],FST_plots[[77]],FST_plots[[73]],FST_plots[[74]],FST_plots[[75]],FST_plots[[76]],FST_plots[[77]],FST_plots[[78]],FST_plots[[79]],FST_plots[[80]],
          FST_plots[[81]],FST_plots[[82]],FST_plots[[88]],FST_plots[[84]],FST_plots[[85]],FST_plots[[86]],FST_plots[[87]],FST_plots[[88]],FST_plots[[89]],FST_plots[[90]],
          FST_plots[[91]],FST_plots[[92]],FST_plots[[93]],FST_plots[[94]],FST_plots[[95]],FST_plots[[96]],FST_plots[[97]],FST_plots[[98]],FST_plots[[99]],FST_plots[[100]],
          nrow=5)

demes <-unique(cg_df$Transplant)
FST_ad_plots <- vector('list', length(demes))
max(cg_df[!is.na(cg_df$FST_sel),]$FST_sel)
min(cg_df[!is.na(cg_df$FST_sel),]$FST_sel)
for (i in seq_along(demes)) {
  dat_filt<-cg_df[cg_df$Transplant%in%demes[i],]
  dat_sum<-cor.test(x=dat_filt$FST_sel, y=dat_filt$Fitness, method = "pearson")
  FST_ad_plots[[i]] <- local({
    i <- i
    p1 <- ggplot(data=dat_filt)+
      geom_smooth(aes(x=FST_sel, y=Fitness), method="lm",se = F)+
      annotate("text",x=0.006,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
      geom_point(aes(x=FST_sel, y=Fitness, col=Home), size=3)+
      labs(shape="Home", colour="Common\nGarden\nLocation")+
      scale_color_manual(values = colfunc(100), guide=F)+
      # ylab("Average relative fitness in common garden")+
      # xlab("Pairwise FST (adapted)")+
      ylab("")+
      xlab("")+
      ylim(0,1)+
      xlim(-0.03,0.06)+
      ggtitle(paste("Transplant to",str_sub(demes[i],2,4)))+
      theme_classic()+
      theme(axis.text=element_text(size=12),axis.title=element_text(size=14)
      )
    print(p1)
  })
}

plot_grid(FST_ad_plots[[1]],FST_ad_plots[[2]],FST_ad_plots[[3]],FST_ad_plots[[4]],FST_ad_plots[[5]],FST_ad_plots[[6]],FST_ad_plots[[7]],FST_ad_plots[[8]],FST_ad_plots[[9]],FST_ad_plots[[10]],
          FST_ad_plots[[11]],FST_ad_plots[[12]],FST_ad_plots[[13]],FST_ad_plots[[14]],FST_ad_plots[[15]],FST_ad_plots[[16]],FST_ad_plots[[17]],FST_ad_plots[[18]],FST_ad_plots[[19]],FST_ad_plots[[20]],
          FST_ad_plots[[21]],FST_ad_plots[[22]],FST_ad_plots[[23]],FST_ad_plots[[24]],FST_ad_plots[[25]],FST_ad_plots[[26]],FST_ad_plots[[27]],FST_ad_plots[[28]],FST_ad_plots[[29]],FST_ad_plots[[30]],
          FST_ad_plots[[31]],FST_ad_plots[[32]],FST_ad_plots[[33]],FST_ad_plots[[34]],FST_ad_plots[[35]],FST_ad_plots[[36]],FST_ad_plots[[37]],FST_ad_plots[[38]],FST_ad_plots[[39]],FST_ad_plots[[40]],
          FST_ad_plots[[41]],FST_ad_plots[[42]],FST_ad_plots[[43]],FST_ad_plots[[44]],FST_ad_plots[[45]],FST_ad_plots[[46]],FST_ad_plots[[47]],FST_ad_plots[[48]],FST_ad_plots[[49]],FST_ad_plots[[50]],
          nrow=5)

plot_grid(FST_ad_plots[[51]],FST_ad_plots[[52]],FST_ad_plots[[53]],FST_ad_plots[[54]],FST_ad_plots[[55]],FST_ad_plots[[56]],FST_ad_plots[[57]],FST_ad_plots[[58]],FST_ad_plots[[59]],FST_ad_plots[[60]],
          FST_ad_plots[[66]],FST_ad_plots[[62]],FST_ad_plots[[63]],FST_ad_plots[[64]],FST_ad_plots[[65]],FST_ad_plots[[66]],FST_ad_plots[[67]],FST_ad_plots[[68]],FST_ad_plots[[69]],FST_ad_plots[[70]],
          FST_ad_plots[[71]],FST_ad_plots[[77]],FST_ad_plots[[73]],FST_ad_plots[[74]],FST_ad_plots[[75]],FST_ad_plots[[76]],FST_ad_plots[[77]],FST_ad_plots[[78]],FST_ad_plots[[79]],FST_ad_plots[[80]],
          FST_ad_plots[[81]],FST_ad_plots[[82]],FST_ad_plots[[88]],FST_ad_plots[[84]],FST_ad_plots[[85]],FST_ad_plots[[86]],FST_ad_plots[[87]],FST_ad_plots[[88]],FST_ad_plots[[89]],FST_ad_plots[[90]],
          FST_ad_plots[[91]],FST_ad_plots[[92]],FST_ad_plots[[93]],FST_ad_plots[[94]],FST_ad_plots[[95]],FST_ad_plots[[96]],FST_ad_plots[[97]],FST_ad_plots[[98]],FST_ad_plots[[99]],FST_ad_plots[[100]],
          nrow=5)

demes <-unique(cg_df$Transplant)
CI_plots <- vector('list', length(demes))
max(cg_df[!is.na(cg_df$D_CI),]$D_CI)
min(cg_df[!is.na(cg_df$D_CI),]$D_CI)
for (i in seq_along(demes)) {
  dat_filt<-cg_df[cg_df$Transplant%in%demes[i],]
  dat_sum<-cor.test(x=dat_filt$D_CI, y=dat_filt$Fitness, method = "pearson")
  CI_plots[[i]] <- local({
    i <- i
    p1 <- ggplot(data=dat_filt)+
      geom_smooth(aes(x=D_CI, y=Fitness), method="lm",se = F)+
      annotate("text",x=0.025,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
      geom_point(aes(x=D_CI, y=Fitness, col=Home), size=3)+
      labs(shape="Home", colour="Common\nGarden\nLocation")+
      scale_color_manual(values = colfunc(100), guide=F)+
      # ylab("Average relative fitness in common garden")+
      # xlab("GF Offset (genome)")+
      ylab("")+
      xlab("")+
      ylim(0,1)+
      xlim(0,0.05)+
      ggtitle(paste("Transplant to",str_sub(demes[i],2,4)))+
      theme_classic()+
      theme(axis.text=element_text(size=12),axis.title=element_text(size=14)
      )
    print(p1)
  })
}

plot_grid(CI_plots[[1]],CI_plots[[2]],CI_plots[[3]],CI_plots[[4]],CI_plots[[5]],CI_plots[[6]],CI_plots[[7]],CI_plots[[8]],CI_plots[[9]],CI_plots[[10]],
          CI_plots[[11]],CI_plots[[12]],CI_plots[[13]],CI_plots[[14]],CI_plots[[15]],CI_plots[[16]],CI_plots[[17]],CI_plots[[18]],CI_plots[[19]],CI_plots[[20]],
          CI_plots[[21]],CI_plots[[22]],CI_plots[[23]],CI_plots[[24]],CI_plots[[25]],CI_plots[[26]],CI_plots[[27]],CI_plots[[28]],CI_plots[[29]],CI_plots[[30]],
          CI_plots[[31]],CI_plots[[32]],CI_plots[[33]],CI_plots[[34]],CI_plots[[35]],CI_plots[[36]],CI_plots[[37]],CI_plots[[38]],CI_plots[[39]],CI_plots[[40]],
          CI_plots[[41]],CI_plots[[42]],CI_plots[[43]],CI_plots[[44]],CI_plots[[45]],CI_plots[[46]],CI_plots[[47]],CI_plots[[48]],CI_plots[[49]],CI_plots[[50]],
          nrow=5)

plot_grid(CI_plots[[51]],CI_plots[[52]],CI_plots[[53]],CI_plots[[54]],CI_plots[[55]],CI_plots[[56]],CI_plots[[57]],CI_plots[[58]],CI_plots[[59]],CI_plots[[60]],
          CI_plots[[66]],CI_plots[[62]],CI_plots[[63]],CI_plots[[64]],CI_plots[[65]],CI_plots[[66]],CI_plots[[67]],CI_plots[[68]],CI_plots[[69]],CI_plots[[70]],
          CI_plots[[71]],CI_plots[[77]],CI_plots[[73]],CI_plots[[74]],CI_plots[[75]],CI_plots[[76]],CI_plots[[77]],CI_plots[[78]],CI_plots[[79]],CI_plots[[80]],
          CI_plots[[81]],CI_plots[[82]],CI_plots[[88]],CI_plots[[84]],CI_plots[[85]],CI_plots[[86]],CI_plots[[87]],CI_plots[[88]],CI_plots[[89]],CI_plots[[90]],
          CI_plots[[91]],CI_plots[[92]],CI_plots[[93]],CI_plots[[94]],CI_plots[[95]],CI_plots[[96]],CI_plots[[97]],CI_plots[[98]],CI_plots[[99]],CI_plots[[100]],
          nrow=5)

CI_ad_plots <- vector('list', length(demes))
max(cg_df[!is.na(cg_df$D_CI_sel),]$D_CI_sel)
min(cg_df[!is.na(cg_df$D_CI_sel),]$D_CI_sel)
for (i in seq_along(demes)) {
  dat_filt<-cg_df[cg_df$Transplant%in%demes[i],]
  dat_sum<-cor.test(x=dat_filt$D_CI_sel, y=dat_filt$Fitness, method = "pearson")
  CI_ad_plots[[i]] <- local({
    i <- i
    p1 <- ggplot(data=dat_filt)+
      geom_smooth(aes(x=D_CI_sel, y=Fitness), method="lm",se = F)+
      annotate("text",x=0.025,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
      geom_point(aes(x=D_CI_sel, y=Fitness, col=Home), size=3)+
      labs(shape="Home", colour="Common\nGarden\nLocation")+
      scale_color_manual(values = colfunc(100), guide=F)+
      # ylab("Average relative fitness in common garden")+
      # xlab("GF Offset (adapted)")+
      ylab("")+
      xlab("")+
      ylim(0,1)+
      xlim(0,0.28)+
      ggtitle(paste("Transplant to",str_sub(demes[i],2,4)))+
      theme_classic()+
      theme(axis.text=element_text(size=12),axis.title=element_text(size=14)
      )
    print(p1)
  })
}

plot_grid(CI_ad_plots[[1]],CI_ad_plots[[2]],CI_ad_plots[[3]],CI_ad_plots[[4]],CI_ad_plots[[5]],CI_ad_plots[[6]],CI_ad_plots[[7]],CI_ad_plots[[8]],CI_ad_plots[[9]],CI_ad_plots[[10]],
          CI_ad_plots[[11]],CI_ad_plots[[12]],CI_ad_plots[[13]],CI_ad_plots[[14]],CI_ad_plots[[15]],CI_ad_plots[[16]],CI_ad_plots[[17]],CI_ad_plots[[18]],CI_ad_plots[[19]],CI_ad_plots[[20]],
          CI_ad_plots[[21]],CI_ad_plots[[22]],CI_ad_plots[[23]],CI_ad_plots[[24]],CI_ad_plots[[25]],CI_ad_plots[[26]],CI_ad_plots[[27]],CI_ad_plots[[28]],CI_ad_plots[[29]],CI_ad_plots[[30]],
          CI_ad_plots[[31]],CI_ad_plots[[32]],CI_ad_plots[[33]],CI_ad_plots[[34]],CI_ad_plots[[35]],CI_ad_plots[[36]],CI_ad_plots[[37]],CI_ad_plots[[38]],CI_ad_plots[[39]],CI_ad_plots[[40]],
          CI_ad_plots[[41]],CI_ad_plots[[42]],CI_ad_plots[[43]],CI_ad_plots[[44]],CI_ad_plots[[45]],CI_ad_plots[[46]],CI_ad_plots[[47]],CI_ad_plots[[48]],CI_ad_plots[[49]],CI_ad_plots[[50]],
          nrow=5)

plot_grid(CI_ad_plots[[51]],CI_ad_plots[[52]],CI_ad_plots[[53]],CI_ad_plots[[54]],CI_ad_plots[[55]],CI_ad_plots[[56]],CI_ad_plots[[57]],CI_ad_plots[[58]],CI_ad_plots[[59]],CI_ad_plots[[60]],
          CI_ad_plots[[66]],CI_ad_plots[[62]],CI_ad_plots[[63]],CI_ad_plots[[64]],CI_ad_plots[[65]],CI_ad_plots[[66]],CI_ad_plots[[67]],CI_ad_plots[[68]],CI_ad_plots[[69]],CI_ad_plots[[70]],
          CI_ad_plots[[71]],CI_ad_plots[[77]],CI_ad_plots[[73]],CI_ad_plots[[74]],CI_ad_plots[[75]],CI_ad_plots[[76]],CI_ad_plots[[77]],CI_ad_plots[[78]],CI_ad_plots[[79]],CI_ad_plots[[80]],
          CI_ad_plots[[81]],CI_ad_plots[[82]],CI_ad_plots[[88]],CI_ad_plots[[84]],CI_ad_plots[[85]],CI_ad_plots[[86]],CI_ad_plots[[87]],CI_ad_plots[[88]],CI_ad_plots[[89]],CI_ad_plots[[90]],
          CI_ad_plots[[91]],CI_ad_plots[[92]],CI_ad_plots[[93]],CI_ad_plots[[94]],CI_ad_plots[[95]],CI_ad_plots[[96]],CI_ad_plots[[97]],CI_ad_plots[[98]],CI_ad_plots[[99]],CI_ad_plots[[100]],
          nrow=5)


demes <-unique(cg_df$Transplant)
DI_plots <- vector('list', length(demes))
max(cg_df[!is.na(cg_df$dM),]$dM)
min(cg_df[!is.na(cg_df$dM),]$dM)
for (i in seq_along(demes)) {
  dat_filt<-cg_df[cg_df$Transplant%in%demes[i],]
  dat_sum<-cor.test(x=dat_filt$dM, y=dat_filt$Fitness, method = "pearson")
  DI_plots[[i]] <- local({
    i <- i
    p1 <- ggplot(data=dat_filt)+
      geom_smooth(aes(x=dM, y=Fitness), method="lm",se = F)+
      annotate("text",x=2,y=0.9,label=paste("r = ",round(dat_sum$estimate[[1]],3),"\np-value = ", signif(dat_sum$p.value,3),sep=""))+
      geom_point(aes(x=dM, y=Fitness, col=Home), size=3)+
      labs(shape="Home", colour="Common\nGarden\nLocation")+
      scale_color_manual(values = colfunc(100), guide=F)+
      #ylab("Average relative fitness in common garden")+
      #xlab("Environmental distance (euclidean)")+
      ylab("")+
      xlab("")+
      ylim(0,1)+
      xlim(0,4)+
      ggtitle(paste("Transplant to",str_sub(demes[i],2,4)))+
      theme_classic()+
      theme(axis.text=element_text(size=12),axis.title=element_text(size=14)
      )
    print(p1)
  })
}

plot_grid(DI_plots[[1]],DI_plots[[2]],DI_plots[[3]],DI_plots[[4]],DI_plots[[5]],DI_plots[[6]],DI_plots[[7]],DI_plots[[8]],DI_plots[[9]],DI_plots[[10]],
          DI_plots[[11]],DI_plots[[12]],DI_plots[[13]],DI_plots[[14]],DI_plots[[15]],DI_plots[[16]],DI_plots[[17]],DI_plots[[18]],DI_plots[[19]],DI_plots[[20]],
          DI_plots[[21]],DI_plots[[22]],DI_plots[[23]],DI_plots[[24]],DI_plots[[25]],DI_plots[[26]],DI_plots[[27]],DI_plots[[28]],DI_plots[[29]],DI_plots[[30]],
          DI_plots[[31]],DI_plots[[32]],DI_plots[[33]],DI_plots[[34]],DI_plots[[35]],DI_plots[[36]],DI_plots[[37]],DI_plots[[38]],DI_plots[[39]],DI_plots[[40]],
          DI_plots[[41]],DI_plots[[42]],DI_plots[[43]],DI_plots[[44]],DI_plots[[45]],DI_plots[[46]],DI_plots[[47]],DI_plots[[48]],DI_plots[[49]],DI_plots[[50]],
          nrow=5)

plot_grid(DI_plots[[51]],DI_plots[[52]],DI_plots[[53]],DI_plots[[54]],DI_plots[[55]],DI_plots[[56]],DI_plots[[57]],DI_plots[[58]],DI_plots[[59]],DI_plots[[60]],
          DI_plots[[66]],DI_plots[[62]],DI_plots[[63]],DI_plots[[64]],DI_plots[[65]],DI_plots[[66]],DI_plots[[67]],DI_plots[[68]],DI_plots[[69]],DI_plots[[70]],
          DI_plots[[71]],DI_plots[[77]],DI_plots[[73]],DI_plots[[74]],DI_plots[[75]],DI_plots[[76]],DI_plots[[77]],DI_plots[[78]],DI_plots[[79]],DI_plots[[80]],
          DI_plots[[81]],DI_plots[[82]],DI_plots[[88]],DI_plots[[84]],DI_plots[[85]],DI_plots[[86]],DI_plots[[87]],DI_plots[[88]],DI_plots[[89]],DI_plots[[90]],
          DI_plots[[91]],DI_plots[[92]],DI_plots[[93]],DI_plots[[94]],DI_plots[[95]],DI_plots[[96]],DI_plots[[97]],DI_plots[[98]],DI_plots[[99]],DI_plots[[100]],
          nrow=5)


cg_ind<-read.table(paste("results/SLiM_output/CG_files/",seed,"_fitnessmat_ind.txt",sep=""),header=T)

colnames(cg_ind)<-rep(paste("ID",seq(1:10000),sep=""))
cg_ind$Transplant<-paste("T",seq(1:100),sep="")

cg_df_ind <- melt(cg_ind,  id.vars ="Transplant",  variable.name = "Individual",
                  value.name = "Fitness")

cg_df_ind$Transplant<-factor(cg_df_ind$Transplant,levels=unique(cg_df_ind$Transplant))

Home<-NULL
for(i in 1:100){
  Home<-c(Home,rep(paste("H",i,sep=""),10000))
}
cg_df_ind$Home <- Home
cg_df_ind$Home<-factor(cg_df_ind$Home,levels=unique(cg_df_ind$Home))


CI<-NULL
for(i in 1:nrow(GF_dist_bef)){
  for(j in 1:nrow(GF_dist_bef)){
    CI<-c(CI,dist(rbind(GF_dist_bef[i,],GF_dist_bef[j,])))
    #CI<-c(CI,CI_bf[j]-CI_bf[i])
  }
}

CI_ind_rep<-NULL
for (i in seq(1,10000,100)){
  CI_ind_rep<-c(CI_ind_rep,rep(CI[i:(i+99)],100))
}

cg_df_ind$D_CI<-CI_ind_rep

GFsel_dist_bef<-data.frame(cbind(gfTrans1_sel1,gfTrans1_sel2))
colnames(GFsel_dist_bef)<-c("CI_env1", "CI_env2")

CI_sel<-NULL
for(i in 1:nrow(GFsel_dist_bef)){
  for(j in 1:nrow(GFsel_dist_bef)){
    CI_sel<-c(CI_sel,dist(rbind(GFsel_dist_bef[i,],GFsel_dist_bef[j,])))
    #CI<-c(CI,CI_bf[j]-CI_bf[i])
  }
}

CI_sel_ind_rep<-NULL
for (i in seq(1,10000,100)){
  CI_sel_ind_rep<-c(CI_sel_ind_rep,rep(CI_sel[i:(i+99)],100))
}

cg_df_ind$D_CI_sel<-CI_sel_ind_rep


CHR_data <- data.frame (SNP = Pfst_pre_filt$LocusName , CHR=Linkage, BP= position1_filt_scaled, P=Pfst_pre_filt$FST)
# Plot !
manhattan(CHR_data , ylim=c(0,0.2),suggestiveline = F, genomewideline = F , logp=F, col=brewer.pal(5, "Set2") , main=plotTitle, ylab=expression(paste("Gen. offset (F"[ST],")")))

sel_pos<-position1_filt[position1_filt%in%unique(pos_T1)]

#FST outlier detection
Pfst_pre_filt_sel<-Pfst_pre_filt[which(Pfst_pre_filt$LocusName%in%select_pos),]
points(x=sel_pos, y=Pfst_pre_filt_sel$FST,col="red", pch=0.7)

#seed = 5003617403563 #ML cline
#seed = 1594110774619 #ML mountain
#seed = 4438659167382 #Case 3
seed = 6698486070847 #Case4

ind<-read.table(paste("results/SLiM_output/",seed,"_ind.txt",sep=""),sep = " ", header=T)

A <- ggplot(data=ind)+
  geom_point(aes(x=opt0, y=phen0))+
  geom_smooth(aes(x=opt0, y=phen0), method="lm",se = T)+
  xlab("Environment 1")+
  ylab("Phenotype 1")+
  ylim(-3,3)+
  theme_classic()

B <- ggplot(data=ind)+
  geom_point(aes(x=opt1, y=phen1))+
  geom_smooth(aes(x=opt0, y=phen0), method="lm",se = T)+
  xlab("Environment 2")+
  ylab("Phenotype 2")+
  ylim(-3,3)+
  theme_classic()

title<- ggdraw() + draw_label(plotTitle, fontface='bold')
plot_grid(title,NULL,A,B,nrow=2,ncol=2,rel_heights = c(1,3))

cor(ind$phen0,ind$opt0)
cor(ind$phen1,ind$opt1)
var(ind$phen0)
var(ind$phen1)

for(i in unique(ind$opt0)){
  print(var(ind[ind$opt0==i,]$phen0))
}

for(i in unique(ind$opt1)){
  print(var(ind[ind$opt1==i,]$phen1))
}


#Individual adapted allele CI values

# top_al_sel1<-unique(cImp1_sel[order(cImp1_sel$y,decreasing = T),]$allele)[1:10]
# length(unique(cImp1_sel$allele))
# A <- ggplot() +
#   geom_line(data=cImp1_sel, aes(x=x , y=y, group=allele),color="red",lwd=1.5) +
#   ylab("Cumulative Importance")+
#   xlab("Environment 1") +
#   theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
#   theme_bw() +
#   theme(axis.text.x = element_text(size = 18, colour = "grey60"),
#         axis.title.x = element_text(size=24)) +
#   theme(axis.text.y = element_text(size = 16, colour = "grey60"),
#         axis.title.y = element_text(size=24, vjust=1)) +
#   theme(strip.text = element_text(size=16)) +
#   geom_text(aes(x=rep(-0.75,10),y= c(seq(0.5,0.3,length.out =10)), label =top_al1))+
#   #annotate("text",x = -1, y =0.5, label=paste(top_al_sel1))+
#   ylim(0.0,0.5)
# 
# top_al2<-unique(cImp2_sel[order(cImp2_sel$y,decreasing = T),]$allele)[1:10]
# length(unique(cImp2_sel$allele))
# B <- ggplot() +
#   geom_line(data=cImp2_sel, aes(x=x , y=y, group=allele),color="red",lwd=1.5) +
#   ylab("Cumulative Importance")+
#   xlab("Environment 2") +
#   theme(plot.margin = unit(c(1.25,1.25,1.25,1.25), "in")) +
#   theme_bw() +
#   theme(axis.text.x = element_text(size = 18, colour = "grey60"),
#         axis.title.x = element_text(size=24)) +
#   theme(axis.text.y = element_text(size = 16, colour = "grey60"),
#         axis.title.y = element_text(size=24, vjust=1)) +
#   theme(strip.text = element_text(size=16)) +
#   #geom_text(aes(x=rep(-0.75,10),y= c(seq(0.5,0.3,length.out =10)), label =top_al_sel2))+
#   ylim(0.0,0.5)
# 
# title <- ggdraw() + draw_label(plotTitle, fontface='bold')
# plot_grid(title,NULL,A,B,nrow=2,ncol=2,rel_heights = c(1,3))

##################################################################
# Common garden visualizations
##################################################################
# SIGMA_K = 0.5
# SIGMA_STAT = 4
# BURNIN_1 = 10000
# BURNIN_2 = 12000
# generation = 10000:12000
#
# sigmak <- NULL
# for(i in 1:length(generation)){
#   sigmak <- c(sigmak,(SIGMA_K * ((generation[i]- BURNIN_1) / (BURNIN_2 - BURNIN_1)) + SIGMA_STAT * (1.0 - (generation[i]- BURNIN_1) / (BURNIN_2 - BURNIN_1))))
# }
# plot(generation,sigmak)

cg_sum<-read.table(paste("results/SLiM_output/CG_files/",seed,"_ML_WF_CG_sum_Gen.txt",sep=""),header=T)
cg_sum[cg_sum$generation==3000,]$local_adaptation

ggplot(data=cg_sum)+
  #geom_line(aes(y=sympatry, x=generation), col="green")+
  #geom_line(aes(y=allopatry, x=generation), col="orange")+
  geom_line(aes(y=local_adaptation, x=generation), col="cyan", lwd=2)+
  geom_vline(xintercept = 1000, linetype='dotted')+
  geom_vline(xintercept = 2000, linetype='dotdash')+
  geom_vline(xintercept = 3000, linetype='dashed')+
  xlab("Generations")+
  ylab("Local Adaptation")+
  annotate("text", y=-0.01, x=500, label= "Burnin", size=2.75)+
  annotate("text", y=-0.01, x=1500, label= "Env.\nTrans.", size=2.75)+
  annotate("text", y=-0.01, x=2200, label= "Adaptation", size=2.75)+
  annotate("text", y=-0.01, x=3200, label= "Climate\nChange", size=2.75)+
  ggtitle(paste(plotTitle)) +
  theme_classic()


ggplot(data=cg_sum)+
  #geom_line(aes(y=sympatry, x=generation), col="green")+
  #geom_line(aes(y=allopatry, x=generation), col="orange")+
  geom_line(aes(y=corr_pheno0_opt0, x=generation), col="orange", lwd=2)+
  geom_line(aes(y=corr_pheno1_opt1, x=generation), col="cyan", lwd=2)+
  geom_vline(xintercept = 1000, linetype='dotted')+
  geom_vline(xintercept = 2000, linetype='dotdash')+
  geom_vline(xintercept = 3000, linetype='dashed')+
  ylim(-1,1)+
  xlab("Generations")+
  ylab("Correlation of phenotypes to optima")+
  annotate("text", y=-1, x=500, label= "Burnin", size=2.75)+
  annotate("text", y=-1, x=1500, label= "Env.\nTrans.", size=2.75)+
  annotate("text", y=-1, x=2200, label= "Adaptation", size=2.75)+
  annotate("text", y=-1, x=3200, label= "Climate\nChange", size=2.75)+
  ggtitle(paste(plotTitle)) +
  theme_classic()


CI_bf1 <- gfTrans1e1$C.Imp_genome_envPop1
CI_bf2 <- gfTrans1e2$C.Imp_genome_envPop2
CI_bf3 <- gfTrans1e3$C.Imp_genome_fakeEnv1
CI_bf4 <- gfTrans1e3$C.Imp_genome_fakeEnv1


ggplot() +
  geom_line(aes(x=1:100, y=GF_dist_bef$CI_env1) ,col="orange", lty=2,size=1.5)+
  geom_line(aes(x=1:100, y=GF_dist_bef$CI_env2) ,col="cyan", lty=2,size=1.5)+
  geom_line(aes(x=1:100, y=GF_dist_bef$CI_fakeEnv1) ,col="red", lty=2,size=1.5)+
  geom_line(aes(x=1:100, y=GF_dist_bef$CI_fakeEnv2) ,col="blue", lty=2,size=1.5)+
  xlab("Location")+
  xlab("Location")+
  ylab("Cumulative Importance")

#case1 <- read.csv("results/R_results/Case1_2889863491989_corr_allEnv.csv")

list.files("results/R_results/")[grep("Case1_",list.files("results/R_results/"))]

case1a <- read.csv("results/R_results/Case1_2889863491989_corr_allEnv.csv")
case1b <- read.csv("results/R_results/Case1_1802264053141_corr_allEnv.csv")
case1c <- read.csv("results/R_results/Case1_1810467011062_corr_allEnv.csv")
case1d <- read.csv("results/R_results/Case1_4196126074995_corr_allEnv.csv")
case1e <- read.csv("results/R_results/Case1_4592125687282_corr_allEnv.csv")
case1f <- read.csv("results/R_results/Case1_1442299973452_corr_allEnv.csv")
case1g <- read.csv("results/R_results/Case1_4128918927628_corr_allEnv.csv")
case1h <- read.csv("results/R_results/Case1_8266232826979_corr_allEnv.csv")
case1i <- read.csv("results/R_results/Case1_8859898358582_corr_allEnv.csv")
case1j <- read.csv("results/R_results/Case1_9698520718839_corr_allEnv.csv")

case1 <- rbind(case1a,case1b,case1c,case1d,case1e,case1f,case1g, case1h, case1i, case1j)

list.files("results/R_results/")[grep("Case2_",list.files("results/R_results/"))]

#case2 <- read.csv("results/R_results/Case2_4207685783970_corr_allEnv.csv")
case2a <- read.csv("results/R_results/Case2_4207685783970_corr_allEnv.csv")
case2b <- read.csv("results/R_results/Case2_1749818164389_corr_allEnv.csv")
case2c <- read.csv("results/R_results/Case2_1688198020681_corr_allEnv.csv")
case2d <- read.csv("results/R_results/Case2_2213962186360_corr_allEnv.csv")
case2e <- read.csv("results/R_results/Case2_4223012060392_corr_allEnv.csv")
case2f <- read.csv("results/R_results/Case2_6141913321101_corr_allEnv.csv")
case2g <- read.csv("results/R_results/Case2_9278521810890_corr_allEnv.csv")
case2h <- read.csv("results/R_results/Case2_6575058632296_corr_allEnv.csv")
case2i <- read.csv("results/R_results/Case2_8134827591478_corr_allEnv.csv")
case2j <- read.csv("results/R_results/Case2_8676247830036_corr_allEnv.csv")

case2 <- rbind(case2a,case2b,case2c,case2d,case2e,case2f,case2g, case2h, case2i, case2j)

#case3 <- read.csv("results/R_results/Case3_1881494860397_corr_allEnv.csv")

list.files("results/R_results/")[grep("Case3_",list.files("results/R_results/"))]

case3a <- read.csv("results/R_results/Case3_1881494860397_corr_allEnv.csv")
case3b <- read.csv("results/R_results/Case3_1292720835655_corr_allEnv.csv")
case3c <- read.csv("results/R_results/Case3_1582596794702_corr_allEnv.csv")
case3d <- read.csv("results/R_results/Case3_2602608872577_corr_allEnv.csv")
case3e <- read.csv("results/R_results/Case3_6099690822883_corr_allEnv.csv")
case3f <- read.csv("results/R_results/Case3_6657938631018_corr_allEnv.csv")
case3g <- read.csv("results/R_results/Case3_5295954239088_corr_allEnv.csv")
case3h <- read.csv("results/R_results/Case3_8792449324158_corr_allEnv.csv")
case3i <- read.csv("results/R_results/Case3_8999846339690_corr_allEnv.csv")
case3j <- read.csv("results/R_results/Case3_9081900117452_corr_allEnv.csv")

case3 <- rbind(case3a,case3b,case3c,case3d,case3e,case3f,case3g, case3h, case3i, case3j)

#case4 <- read.csv("results/R_results/Case4_4693516184575_corr_allEnv.csv")

list.files("results/R_results/")[grep("Case4_",list.files("results/R_results/"))]

case4a <- read.csv("results/R_results/Case4_4693516184575_corr_allEnv.csv")
case4b <- read.csv("results/R_results/Case4_6153876761206_corr_allEnv.csv")
case4c <- read.csv("results/R_results/Case4_1723367103375_corr_allEnv.csv")
case4d <- read.csv("results/R_results/Case4_2591104706283_corr_allEnv.csv")
case4e <- read.csv("results/R_results/Case4_4030688734724_corr_allEnv.csv")
case4f <- read.csv("results/R_results/Case4_5434225959004_corr_allEnv.csv")
case4g <- read.csv("results/R_results/Case4_6004202094859_corr_allEnv.csv")
case4h <- read.csv("results/R_results/Case4_8435562691650_corr_allEnv.csv")
case4i <- read.csv("results/R_results/Case4_8748231054981_corr_allEnv.csv")
case4j <- read.csv("results/R_results/Case4_9785941433393_corr_allEnv.csv")

case4 <- rbind(case4a,case4b,case4c,case4d,case4e,case4f,case4g, case4h, case4i, case4j)

x <- rbind(case1,case2,case3,case4)
x$case <- c(rep("case 1",nrow(case1)), rep("case 2",nrow(case2)), rep("case 3",nrow(case3)), rep("case 4",nrow(case4)))

#x <- rbind(case1a,case2a,case3a,case4a)
#x$case <- c("case 1", "case 2","case 3","case 4")

#x_mean <- colMeans(x[,-25])
#x_sd <- sapply(x[,-25], function(x) sd(x))

y<-melt(x)

require(tidyverse)
z <- y %>%
  group_by(case, variable)%>%
  summarise (mean = mean(value), IQR = IQR(value), TwoFive = quantile(value,0.25), SevenFive = quantile(value,0.75)) 
  
# ggplot(data=x)+
#   geom_col(aes(x=case, y=GF_off_genome_allEnv_Edges))+
#   geom_col(aes(x=case, y=GF_off_genome_allEnv_Cores))
# 
# ggplot(data=x)+
#   geom_col(aes(x=case, y=GF_off_genome_causEnv_Edges))+
#   geom_col(aes(x=case, y=GF_off_genome_causEnv_Cores))

# y_edge <- y[grep("Edge",y$variable),]
# y_core <- y[grep("Core",y$variable),]

z_edge <- z[grep("Edge",z$variable),]
z_core <- z[grep("Core",z$variable),]

z_edge <- z_edge[z_edge$variable!="Env_ED_causP2_env_Edges",]
z_edge <- z_edge[z_edge$variable!="Env_MD_causP2_env_Edges",]
z_core <- z_core[z_core$variable!="Env_ED_causP2_env_Cores",]
z_core <- z_core[z_core$variable!="Env_MD_causP2_env_Cores",]

E1<-ggplot(z_edge[z_edge$case=="case 1",], aes(variable, mean, fill=variable)) + 
  geom_bar(position="dodge",stat="identity")+
  geom_errorbar(aes(ymin=TwoFive, ymax=SevenFive), width=.2,
                position=position_dodge(.9))+
  theme_classic()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(axis.title.y=element_blank())+ 
  theme(legend.position = "none")+ 
  theme(axis.text.y = element_text(size = 20))+
  scale_y_continuous(trans = "reverse")+
  ggtitle("Case 1 Edge")+
  ylim(0.1,-1)
  
C1<-ggplot(z_core[z_core$case=="case 1",], aes(variable, mean, fill=variable)) + 
  geom_bar(position="dodge",stat="identity")+
  geom_errorbar(aes(ymin=TwoFive, ymax=SevenFive), width=.2,
                position=position_dodge(.9))+
  theme_classic()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(legend.position = "none")+ 
  theme(axis.text.y = element_text(size = 20))+
  scale_y_continuous(trans = "reverse")+
  ylim(0.1,-1)+
  ggtitle("Case 1 Core")+
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())

E2<-ggplot(z_edge[z_edge$case=="case 2",], aes(variable, mean, fill=variable)) + 
  geom_bar(position="dodge",stat="identity")+
  geom_errorbar(aes(ymin=TwoFive, ymax=SevenFive), width=.2,
                position=position_dodge(.9))+
  theme_classic()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(axis.title.y=element_blank())+ 
  theme(legend.position = "none")+ 
  theme(axis.text.y = element_text(size = 20))+
  scale_y_continuous(trans = "reverse")+
  ggtitle("Case 2 Edge")+
  ylim(0.1,-1)

C2<-ggplot(z_core[z_core$case=="case 2",], aes(variable, mean, fill=variable)) + 
  geom_bar(position="dodge",stat="identity")+
  geom_errorbar(aes(ymin=TwoFive, ymax=SevenFive), width=.2,
                position=position_dodge(.9))+
  theme_classic()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(legend.position = "none")+ 
  theme(axis.text.y = element_text(size = 20))+
  scale_y_continuous(trans = "reverse")+
  ylim(0.1,-1)+
  ggtitle("Case 2 Core")+
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())

E3<-ggplot(z_edge[z_edge$case=="case 3",], aes(variable, mean, fill=variable)) + 
  geom_bar(position="dodge",stat="identity")+
  geom_errorbar(aes(ymin=TwoFive, ymax=SevenFive), width=.2,
                position=position_dodge(.9))+
  theme_classic()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(axis.title.y=element_blank())+ 
  theme(legend.position = "none")+ 
  theme(axis.text.y = element_text(size = 20))+
  scale_y_continuous(trans = "reverse")+
  ggtitle("Case 3 Edge")+
  ylim(0.1,-1)

C3<-ggplot(z_core[z_core$case=="case 3",], aes(variable, mean, fill=variable)) + 
  geom_bar(position="dodge",stat="identity")+
  geom_errorbar(aes(ymin=TwoFive, ymax=SevenFive), width=.2,
                position=position_dodge(.9))+
  theme_classic()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(legend.position = "none")+ 
  theme(axis.text.y = element_text(size = 20))+
  scale_y_continuous(trans = "reverse")+
  ggtitle("Case 3 Core")+
  ylim(0.1,-1)+
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())

E4<-ggplot(z_edge[z_edge$case=="case 4",], aes(variable, mean, fill=variable)) + 
  geom_bar(position="dodge",stat="identity")+
  geom_errorbar(aes(ymin=TwoFive, ymax=SevenFive), width=.2,
                position=position_dodge(.9))+
  theme_classic()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(axis.title.y=element_blank())+ 
  theme(legend.position = "none")+ 
  theme(axis.text.y = element_text(size = 20))+
  scale_y_continuous(trans = "reverse")+
  ggtitle("Case 4 Edge")+
  ylim(0.1,-1)

C4<-ggplot(z_core[z_core$case=="case 4",], aes(variable, mean, fill=variable)) + 
  geom_bar(position="dodge",stat="identity")+
  geom_errorbar(aes(ymin=TwoFive, ymax=SevenFive), width=.2,
                position=position_dodge(.9))+
  theme_classic()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(legend.position = "none")+ 
  theme(axis.text.y = element_text(size = 20))+
  scale_y_continuous(trans = "reverse")+
  ylim(0.1,-1)+
  ggtitle("Case 4 Core")+
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())+
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())

E1l<-ggplot(z_edge[z_edge$case=="case 1",], aes(variable, mean, fill=variable)) + 
  geom_bar(position="dodge",stat="identity")+
  geom_errorbar(aes(ymin=TwoFive, ymax=SevenFive), width=.2,
                position=position_dodge(.9))+
  theme_classic()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  theme(axis.title.y=element_blank())+ 
  scale_fill_discrete("", labels=c('GF off. genome, all env.', 'GF off. genome, causal env.', 'GF off. causal, all env.', 'GF off. causal, causal env.', 'Causal Env. dist. (ED)','Causal Env. dist. (MD)','All Env. dist. (ED)', 'All Env. dist. (MD)', 'FST genome', 'FST causal'))+
  #scale_fill_discrete("", labels=c('GF off. genome, all env.', 'GF off. genome, causal env.', 'GF off. causal, all env.', 'GF off. causal, causal env.', 'Causal Env. dist. (ED)','Causal Env. dist. (MD)','All Env. dist. (ED)', 'All Env. dist. (MD)','Causal Env. plus 2 dist. (ED)', 'Causal Env. plus 2 dist. (MD)', 'FST genome', 'FST causal'))+ 
  theme(axis.text.y = element_text(size = 20))+
  scale_y_continuous(trans = "reverse")+
  ylim(0.1,-1)

leg<-get_legend(E1l+theme(legend.margin=margin(t=5, r=0, b=0, l=0, unit="cm")))

plot_grid(E1,C1,NULL,E2,C2,leg,E3,C3,NULL, E4,C4,NULL,ncol=3, rel_widths = c(1,0.9,0.6))


plot(Popsenv$fakeEnv2,type="l", ylab="")
lines(Popsenv$fakeEnv1, col="grey")
lines(Popsenv$envPop1, col="red")
lines(Popsenv$envPop2, col="blue")


Case1 <- c(0.28721,0.2834,0.285939,0.285249,0.285564,0.287348,0.281622,0.284823,0.28792,0.288161)
Case2 <- c(0.237691,0.241765,0.238952,0.23543,0.236615,0.236157,0.234768,0.233871,0.235267,0.24091)
Case3 <- c(0.173907,0.171456,0.172676,0.175239,0.169022,0.169979,0.1759,0.170812,0.170504,0.169133)
Case4 <- c(0.206885,0.203676,0.206336,0.203341,0.208968,0.205407,0.20297,0.207721,0.206723,0.204104)

LA <- cbind(Case1, Case2, Case3, Case4)


LA_summary <- melt(LA) %>% # the names of the new data frame and the data frame to be summarised
  group_by(Var2) %>%   # the grouping variable
  summarise(mean_LA = mean(value),  # calculates the mean of each group
            sd_LA = sd(value), # calculates the standard deviation of each group
            n_LA = n(),  # calculates the sample size per group
            SE_LA = sd(value)/sqrt(n())) # calculates the standard error of each group

LA_Plot <- ggplot(LA_summary, aes(Var2, mean_LA)) + 
  geom_col() +  
  geom_errorbar(aes(ymin = mean_LA - sd_LA, ymax = mean_LA + sd_LA), width=0.2)

LA_Plot + labs(y="Local Adaptation  s.d.", x = "") + theme_classic()
